<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: pestfind
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="index.html" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">

pestfind
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<!--END OF HEADER-->






<H2>
    Function
</H2>
Finds PEST motifs as potential proteolytic cleavage sites

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
  -->

<b>pestfind</b> finds PEST domains in a protein sequence.

<p>

Proteins with intracellular half-lives of less than two hours are found
to contain regions rich in proline, glutamic acid, serine and threonine
(P, E, S and T).  These so called PEST regions are generally flanked by
clusters of positively charged amino acids. 

<p>

<b>pestfind</b> identifies possible PEST regions in a protein sequence
using the Molecular fraction of the P, E, S and T components, and
the hydrophobicity index of the region. 













<H2>
    Algorithm
</H2>

<!-- 
        Algorithms
	  Pseudocode/Vague description/Handwaving/web links to useful pages
  -->

 The algorithm comprises first selecting all
 portions of a submitted sequence which correspond to the following
 Perl regular expression:<P>

  <CENTER><B>/([HKRhkr]|^\w)(\w*?)([HKRhkr]|\w$)/</B></CENTER><P>

  Or in English, the following sequence of conditions:<P>

<UL>
<LI> Either one of the residues H, K and R or the first residue in the
     sequence
<LI> The shortest sequence of residues excluding H, K and R
<LI> Either one of the residues H, K and R or the last residue in the
     sequence
</UL>

  This isolates all of the initial candidate sequences.  Following this
 selection, a set of inflexible rules are applied.  First, the candidate
 sequence must be of at least the length specified at execution (default is
 10 AAs).  Next, the sequence must conform to the following Perl regular
 expression:<P>

  <CENTER><B>/P/ && /(?:D|E)/ && /(?:S|T)/</B></CENTER><P>

  Or in English, the following sequence of conditions:<P>

<UL>
<LI> Any residues including at least one P
<LI> Any residues including at least either one D or one E
<LI> Any residues including at least either one S or one T
</UL>

  Once these conditions are met, we are left with a set of all candidate
 sequences from which we select the suitable members using the PEST SCORE
 algorithm.  Since this gives us an arbitary cut off point, it is possible
 to ask for all candidate sequences with their PEST scores.  PEST Scores
 are derived from two factors, the Molecular Fraction of the D, P, E, S and
 T components of the candidate and it's hydrophobic index.<P>

  The PEDST molecular fraction is given by the following equation: <P>

  <CENTER><B>PEDST = 100 * (S - E)  / T</B></CENTER><P>

  S = Total molecular weights of the P, E, D, S and T components of the
      candidate sequence<BR>
  E = Sum of the molecular weights of a single E, P and T Molecule (327)<BR>
  T = Total molecular weights of all amino acids within the candidate
      sequence <P>

  The Hydrophobic index is given by the following equation: <P>

  <CENTER><B>HI = sum x=aa {H(x) * F(x)}</B></CENTER><P>

  H(x) = Hydrophobicity of amino acid (x)<BR>
  F(x) = Molar fraction of amino acid (x) within candidate sequence<P>

  The PEST score is calculated from the PEDST molar fraction and the
 hydrophobic index from the following equation: <P>

  <CENTER><B>PEST Score = -1 * ((HI * 0.5) - (PEDST * 0.55))</B></CENTER><P>

  A score in excess of the cut-off is recorded as a potential PEST
 sequence, so a low hydrophobicity index and a high PEDST score is most
 desirable.<P>













<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with pestfind</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>pestfind -graph cps </b>
Finds PEST motifs as potential proteolytic cleavage sites
Enter input sequence: <b>../../data/exu2_drops.embl</b>
Enter window length [10]: <b></b>
Sort order of results
      1 : length
      2 : position
      3 : score
Select sort order of results [score]: <b></b>
Enter output filename [exu2_drops.pestfind]: <b></b>

Created pestfind.ps

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>




<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Mandatory qualifiers:
  [-sequence]          sequence   Protein sequence USA to be analysed.
   -window             integer    Minimal distance between positively charged
                                  amino acids.
   -order              selection  Results may be sorted by length, position
                                  and score.
   -outfile            outfile    Name of file to which results will be
                                  written.
   -graph              xygraph    Graph type

   Optional qualifiers:
   -threshold          float      Threshold value to discriminate weak from
                                  potential PEST motifs.
   -aadata             string     File of amino acid properties and molecular
                                  masses.

   Advanced qualifiers:
   -[no]potential      boolean    Decide whether potential PEST motifs should
                                  be printed.
   -[no]poor           boolean    Decide whether poor PEST motifs should be
                                  printed.
   -invalid            boolean    Decide whether invalid PEST motifs should be
                                  printed.
   -[no]map            boolean    Decide whether PEST motifs should be mapped
                                  to sequence.

   General qualifiers:
  -help                boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose


</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#f5f5ff">
<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Mandatory qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-sequence]<br>(Parameter 1)</td>
<td>Protein sequence USA to be analysed.</td>
<td>Readable sequence</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-window</td>
<td>Minimal distance between positively charged amino acids.</td>
<td>Integer 2 or more</td>
<td>10</td>
</tr>

<tr>
<td>-order</td>
<td>Results may be sorted by length, position and score.</td>
<td>Choose from selection list of values</td>
<td>score</td>
</tr>

<tr>
<td>-outfile</td>
<td>Name of file to which results will be written.</td>
<td>Output file</td>
<td><i>&lt;sequence&gt;</i>.pestfind</td>
</tr>

<tr>
<td>-graph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including postscript, ps, hpgl, hp7470, hp7580, meta, colourps, cps, xwindows, x11, tektronics, tekt, tek4107t, tek, none, null, text, data, xterm, png</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Optional qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-threshold</td>
<td>Threshold value to discriminate weak from potential PEST motifs.</td>
<td>Number from -55.00 to 55.00</td>
<td>+5.0</td>
</tr>

<tr>
<td>-aadata</td>
<td>File of amino acid properties and molecular masses.</td>
<td>Any string is accepted</td>
<td>Eamino.dat</td>
</tr>

<tr bgcolor="#FFFFD0">
<th align="left" colspan=2>Advanced qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-[no]potential</td>
<td>Decide whether potential PEST motifs should be printed.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-[no]poor</td>
<td>Decide whether poor PEST motifs should be printed.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr>
<td>-invalid</td>
<td>Decide whether invalid PEST motifs should be printed.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-[no]map</td>
<td>Decide whether PEST motifs should be mapped to sequence.</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

</table>


<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>pestfind</b> reads any normal sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ../../data/exu2_drops.embl</h3>
<table width="90%"><tr><td bgcolor="#FFDDFF">
<pre>
ID   EXU2_DROPS     STANDARD;      PRT;   477 AA.
AC   Q24617;
DT   01-NOV-1997 (Rel. 35, Created)
DT   01-NOV-1997 (Rel. 35, Last sequence update)
DT   01-NOV-1997 (Rel. 35, Last annotation update)
DE   Maternal exuperantia 2 protein.
GN   EXU2.
OS   Drosophila pseudoobscura (Fruit fly).
OC   Eukaryota; Metazoa; Arthropoda; Tracheata; Hexapoda; Insecta;
OC   Pterygota; Neoptera; Endopterygota; Diptera; Brachycera; Muscomorpha;
OC   Ephydroidea; Drosophilidae; Drosophila.
OX   NCBI_TaxID=7237;
RN   [1]
RP   SEQUENCE FROM N.A.
RX   MEDLINE=94350208; PubMed=8070663;
RA   Luk S.K.-S., Kilpatrick M., Kerr K., Macdonald P.M.;
RT   "Components acting in localization of bicoid mRNA are conserved among
RT   Drosophila species.";
RL   Genetics 137:521-530(1994).
CC   -!- FUNCTION: ENSURES THE PROPER LOCALIZATION OF THE MRNA OF THE
CC       BICOID GENE TO THE ANTERIOR REGIONS OF THE OOCYTE THUS PLAYING
CC       A FUNDAMENTAL ROLE IN THE ESTABLISHMENT OF THE POLARITY OF THE
CC       OOCYTE. MAY BIND THE BCD MRNA (BY SIMILARITY).
CC   --------------------------------------------------------------------------
CC   This SWISS-PROT entry is copyright. It is produced through a collaboration
CC   between  the Swiss Institute of Bioinformatics  and the  EMBL outstation -
CC   the European Bioinformatics Institute.  There are no  restrictions on  its
CC   use  by  non-profit  institutions as long  as its content  is  in  no  way
CC   modified and this statement is not removed.  Usage  by  and for commercial
CC   entities requires a license agreement (See http://www.isb-sib.ch/announce/
CC   or send an email to license@isb-sib.ch).
CC   --------------------------------------------------------------------------
DR   EMBL; L22553; AAA28523.1; -.
DR   FlyBase; FBgn0014783; Dpse\exu2.
KW   Developmental protein; RNA-binding.
FT   DOMAIN      249    271       GLU/SER/PRO/THR-RICH (PEST REGION).
SQ   SEQUENCE   477 AA;  53194 MW;  8B5334A77DB9467B CRC64;
     MVSAISEDSA SATASGQCEV VKEELPAGNY ILVAVEIDTT GRRLIDEIVQ LAGYTSKGNF
     QQYIMPYMNL NQAARQRHQI RVISIGFYRM LKSMQTYKII KSKSEVAALM DFLNWLETLL
     AKQPNKEGIV MLYHDDRKFI PYMILEALKK YSLIDRFNRS VKAFANTCPM AKTFLGKHGI
     KNCGLRKLSM LLAKSKDGNS TKEDEHENPE GNSSITDNSG HKNQKQGAFE GSANVRAKMV
     YEMALQLIES ESTESPESFE SPESSESSEA EVKLLNAVRP FSQLLSSTIL ELKDQNHSLG
     RQNSFRPVFL NYFRTTLNYR VRAVKYRIGL AEHGFTLKSL KAIWSDKRKP GLELVLTAID
     SLKTEETAEL LDLLDSYYDP SKTTIKPRCK RSGNGTRRRN RAKGAASSKN GAIGAGGDNS
     VPDSATKPGG RPRRKRNNIR NNILGPQNTE KGSPKAEMKT STPKSMSIKP PSEFADI
//
</pre>
</td></tr></table><p>






<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
<p>

The output is a standard EMBOSS report file. 

<p>

The results can be output in one of several styles by using the
command-line qualifier <b>-rformat xxx</b>, where 'xxx' is replaced by
the name of the required format.  The available format names are: embl,
genbank, gff, pir, swiss, trace, listfile, dbmotif, diffseq, excel,
feattable, motif, regions, seqtable, simple, srs, table, tagseq

<p>

See:
<A href="http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html">
http://www.uk.embnet.org/Software/EMBOSS/Themes/ReportFormats.html</A>
for further information on report formats.

<p>

   -->

<b>pestfind</b> 
outputs a graph to the specified graphics device. 
outputs a report format file. The default format is ...


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: exu2_drops.pestfind</h3>
<table width="90%"><tr><td bgcolor="#DDFFDD">
<pre>
PEST-find: Finds PEST motifs as potential proteolytic cleavage sites.

     8 PEST motifs were identified in EXU2_DROPS
       from positions 1 to 477 and sorted by score.

Potential PEST motif with 34 amino acids between position 238 and 273.
   238 KMVYEMALQLIESESTESPESFESPESSESSEAEVK 273
       DEPST: 54.20 % (w/w)
       Hydrophobicity index: 38.12
       PEST score: 10.75  

Potential PEST motif with 14 amino acids between position 206 and 221.
   206 HENPEGNSSITDNSGH 221
       DEPST: 35.59 % (w/w)
       Hydrophobicity index: 26.81
       PEST score: 6.17  

Poor PEST motif with 18 amino acids between position 363 and 382.
   363 KTEETAELLDLLDSYYDPSK 382
       PEST score: 2.90  

Poor PEST motif with 17 amino acids between position 409 and 427.
   409 KNGAIGAGGDNSVPDSATK 427
       PEST score: -9.60  

Poor PEST motif with 19 amino acids between position 22 and 42.
    22 KEELPAGNYILVAVEIDTTGR 42
       PEST score: -10.92  

Poor PEST motif with 10 amino acids between position 440 and 451.
   440 RNNILGPQNTEK 451
       PEST score: -15.67  

Poor PEST motif with 13 amino acids between position 279 and 293.
   279 RPFSQLLSSTILELK 293
       PEST score: -16.51  

Poor PEST motif with 13 amino acids between position 349 and 363.
   349 KPGLELVLTAIDSLK 363
       PEST score: -19.67  


       ---------+---------+---------+---------+---------+---------+

     1 MVSAISEDSASATASGQCEVVKEELPAGNYILVAVEIDTTGRRLIDEIVQLAGYTSKGNF 60
                             OOOOOOOOOOOOOOOOOOO                   

    61 QQYIMPYMNLNQAARQRHQIRVISIGFYRMLKSMQTYKIIKSKSEVAALMDFLNWLETLL 120
                                                                   

   121 AKQPNKEGIVMLYHDDRKFIPYMILEALKKYSLIDRFNRSVKAFANTCPMAKTFLGKHGI 180
                                                                   

   181 KNCGLRKLSMLLAKSKDGNSTKEDEHENPEGNSSITDNSGHKNQKQGAFEGSANVRAKMV 240
                                 ++++++++++++++                  ++

   241 YEMALQLIESESTESPESFESPESSESSEAEVKLLNAVRPFSQLLSSTILELKDQNHSLG 300
       ++++++++++++++++++++++++++++++++       OOOOOOOOOOOOO        

   301 RQNSFRPVFLNYFRTTLNYRVRAVKYRIGLAEHGFTLKSLKAIWSDKRKPGLELVLTAID 360
                                                        OOOOOOOOOOO

   361 SLKTEETAELLDLLDSYYDPSKTTIKPRCKRSGNGTRRRNRAKGAASSKNGAIGAGGDNS 420
       OO OOOOOOOOOOOOOOOOOO                            OOOOOOOOOOO

   421 VPDSATKPGGRPRRKRNNIRNNILGPQNTEKGSPKAEMKTSTPKSMSIKPPSEFADI 477
       OOOOOO              OOOOOOOOOO                           

       Symbols   PEST motifs
       +++++++   potential
       OOOOOOO   poor     

</pre>
</td></tr></table><p>
<p><h3>Graphics File: pestfind.ps</h3>
<p><img src="pestfind.1.pestfind.gif" alt="[pestfind results]">






<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
<p>
EMBOSS data files are distributed with the application and stored
in the standard EMBOSS data directory, which is defined
by the EMBOSS environment variable EMBOSS_DATA.

<p>

To see the available EMBOSS data files, run:
<p>
<pre>
% embossdata -showall
</pre>
<p>
To fetch one of the data files (for example 'Exxx.dat') into your
current directory for you to inspect or modify, run:

<pre>

% embossdata -fetch -file Exxx.dat

</pre>
<p>

Users can provide their own data files in their own directories.
Project specific files can be put in the current directory, or for
tidier directory listings in a subdirectory called
".embossdata". Files for all EMBOSS runs can be put in the user's home
directory, or again in a subdirectory called ".embossdata".

<p>
The directories are searched in the following order:

<ul>
   <li> . (your current directory)
   <li> .embossdata (under your current directory)
   <li> ~/ (your home directory)
   <li> ~/.embossdata
</ul>
<p>

   -->

None.














<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr><td><a href="antigenic.html">antigenic</a></td><td>Finds antigenic sites in proteins</td></tr>
<tr><td><a href="digest.html">digest</a></td><td>Protein proteolytic enzyme or reagent cleavage digest</td></tr>
<tr><td><a href="fuzzpro.html">fuzzpro</a></td><td>Protein pattern search</td></tr>
<tr><td><a href="fuzztran.html">fuzztran</a></td><td>Protein pattern search after translation</td></tr>
<tr><td><a href="helixturnhelix.html">helixturnhelix</a></td><td>Report nucleic acid binding motifs</td></tr>
<tr><td><a href="oddcomp.html">oddcomp</a></td><td>Finds protein sequence regions with a biased composition</td></tr>
<tr><td><a href="patmatdb.html">patmatdb</a></td><td>Search a protein sequence with a motif</td></tr>
<tr><td><a href="patmatmotifs.html">patmatmotifs</a></td><td>Search a PROSITE motif database with a protein sequence</td></tr>
<tr><td><a href="pepcoil.html">pepcoil</a></td><td>Predicts coiled coil regions</td></tr>
<tr><td><a href="preg.html">preg</a></td><td>Regular expression search of a protein sequence</td></tr>
<tr><td><a href="pscan.html">pscan</a></td><td>Scans proteins using PRINTS</td></tr>
<tr><td><a href="sigcleave.html">sigcleave</a></td><td>Reports protein signal cleavage sites</td></tr>
</table>

<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>
<!--
        Who has worked on the program in the past.
   -->

This application was written by Michael K. Schuster, 
Department of Medical Biochemistry, University Vienna 
<a href="mailto:martin.grabner@univie.ac.at">(martin.grabner@univie.ac.at)</a>

















<H2>
    History
</H2>

<!--
        What changes have been made.
   -->

This is based on the algorithm used in the QuickBasic program by Rodgers et al.

<p>

Written (November 2002) - Michael K. Schuster















<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.












</BODY>
</HTML>

