<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: DOMAINSEQS documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<!--#include virtual="/apps/header1.inc" -->
<H2> DOMAINSEQS documentation
<!--#include virtual="/apps/header2.inc" -->
<!-- END OF HEADER -->





<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<br><H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>

<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>



<!--#include file="inc/domainseqs.ione" -->



<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
DOMAINSEQS parses a DCF file (domain classification file) and writes a file 
containing the same data, except that domain sequence 
information derived from structural and, optionally, sequence databases are
added.  
Domain sequences are taken from domain CCF files (clean coordinate files) 
and (optionally) the swissprot database.  

If the swissprot sequence is used, DOMAINSEQS requires a swissprot:PDB 
equivalence file that gives the accession number of each PDB file 
corresponding to the domains in the DCF file.  

The path for the CCF files (input) and names of DCF files (input and output) are specified by the user.  A log file is also written.






<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.

     Cannot use the test data files because they might be empty or need 
     hand-editing 

     Use "<b>DOMAINSEQS</b> reads any normal sequence USAs." if
     appropriate.
-->
<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>
The format of the DCF file is described in the <a href="scopparse.html">SCOPPARSE documentation.</a>

<!--#include file="inc/domainseqs.input" -->

<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>
The format of the DCF file is described in the <a href="scopparse.html">SCOPPARSE documentation.</a>
<br>
DOMAINSEQS may add the following records:
<UL>
<LI>(1) AC - Accession number of the domain sequence.  This record will only 
be present if the DCF file has been processed using 
DOMAINSEQS and if an accession number for the PDB file corresponding to the 
domain is given in the swissprot:PDB equivalence file (generated by 
PDBTOSP) that DOMAINSEQS makes use of.
<LI>(2) SP - Swissprot code of the domain sequence.  This record will only 
be present if the domain classification file has been processed using 
DOMAINSEQS and if an swissprot code for the PDB file corresponding to the 
domain is given in the swissprot:PDB equivalence file (generated by 
PDBTOSP) that DOMAINSEQS makes use of.
<LI>(3) RA - Position of domain in swissprot sequence.  The integers 
preceeding START and END give the start and end points respectively of the
domain sequence relative to the full-length swissprot sequence.
<LI>(4) SQ - Sequence of the domain according to swissprot.  This sequence
is taken from the swissprot database.  The SQ record will only be 
present if the SCOP classification file has been processed using 
DOMAINSEQS and if an accession number for the PDB file corresponding to the 
domain is given in the swissprot:PDB equivalence file (generated by 
PDBTOSP) that DOMAINSEQS makes use of.
</UL>


<!--#include file="inc/domainseqs.output" -->

<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>

If the user specified retrieval of sequences from a sequence database, 
DOMAINSEQS uses a swissprot:PDB equivalence file is generated by using 
<a href="pdbtosp.html">PDBTOSP.</a>







<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>
<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 
<pre>
<!--#include file="inc/domainseqs.ihelp" -->
</pre>
</td></tr></table>
<P>
<!--#include file="inc/domainseqs.itable" -->
 


<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of DOMAINSEQS is shown below.

<!--#include file="inc/domainseqs.usage" -->

<br><br>
<!-- The DCF file (domain classification file) lacking sequence information called /test_data/all.scop was parsed and a DCF file with sequence information called /test_data/domainseqs/all_s.scop was written.  The file Epdbtosp.dat was queried for accession numbers of PDB files corresponding to the domains in all.scop, but no accession numbers were found for either of the domains d1ii7a_ or d1cs4a_.  A log file called /test_data/domainseqs/domainseqs.log was written. -->



<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>

The swissprot:PDB equivalence file available as part of the EMBOSS
distribution does not contain accession numbers for all PDB files and covers
only a relatively small proportion of domains in SCOP.




<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>


DOMAINSEQS will not attempt to retrieve a swissprot sequence for any domains 
comprised of more than a single segment, i.e. for domains whose NC record 
in the DCF file has a value other than 1 
(see <a href="scopparse.html">SCOPPARSE documentation</a>).
<br>
The region of sequence to retrieve is identified by alignment of the sequence from the
CCF file (structural database) to the full length sequences in swissprot.   If this were to be done
for segmented domains, the start and end point for the retrieved sequence 
(relative to the full length sequence) might actually include a completely 
different domain.
<br><br>
Consider :
<br> (A) represents a sequence from the coordinate file for a segmented domain D1
<br> (B) represents the full-length swissprot sequence, it includes D1, but D1 is
 split by a second domain D2
<pre> 
                 D1
  (A)  XXXXXXXXXXXXXXXXXXXXX
  
          D1         D2        D1
  (B) XXXXXXXXXX-----------XXXXXXXXXXX
</pre>
  
 It should be clear that for an alignment retrieving B, the start and end 
 points for the retrieved sequence would be misleading.

<br><br>
The user should be aware that sequences from the domain CCF file
for domains comprised of more than a single segment are not biologically 
significant, as the sequences are derived from different segments of one or
more chains.  However, such sequences might be acceptable for redundancy 
calculations (e.g. by using <a href="domainnr.html">DOMAINNR</a>) because 
two redundant domains made of similar fragments will have similar sequences, 
so the redundancy should be detectable.

<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b> Domain classification file (for SCOP)</b></td>
<td> DCF format (EMBL-like format for domain classification data). </td>
<td> Classification and other data for domains from SCOP.  </td>
<td> <a href="scopparse.html">SCOPPARSE</a> </td>
<td> Domain sequence information can be added to the file by using DOMAINSEQS.  </td>
</tr>
<tr>
<td><b> Domain classification file (for CATH)</b></td>
<td> DCF format (EMBL-like format for domain classification data). </td>
<td> Classification and other data for domains from CATH.  </td>
<td> <a href="cathparse.html">CATHPARSE</a> </td>
<td> Domain sequence information can be added to the file by using DOMAINSEQS.  </td>
</tr>
<tr>
<td><b> Clean coordinate file (for protein) </b></td>
<td> CCF format (EMBL-like format for protein coordinate and derived data).</td>
<td> Coordinate and other data for a single PDB file.  The data are 'cleaned-up': self-consistent and error-corrected.  </td>
<td> <a href="pdbparse.html">PDBPARSE</a> </td>
<td> Records for residue solvent accessibility and secondary structure are added 
    to the file by using PDBPLUS. </td>
</tr>
<tr>
<td><b> Clean coordinate file (for domain)</b></td>
<td> CCF format (EMBL-like format for protein coordinate and derived data).</td>
<td> Coordinate and other data for a single domain from SCOP or CATH. The data are 'cleaned-up': self-consistent and error-corrected. </td>
<td> <a href="domainer.html">DOMAINER</a> </td>
<td> Records for residue solvent accessibility and secondary structure are added 
to the file by using PDBPLUS.</td>
</tr>
<tr>
<td><b> swissprot:PDB equivalence file</b></td>
<td> EMBL-like format. </td>
<td> A file containing swissprot identifiers for PDB codes. </td>
<td> Included in the EMBOSS distribution</td>
<td> N.A. </td>
</tr>
</table>





<!--#include file="inc/domainseqs.history" -->
<!--#include file="inc/domainseqs.comment" -->


<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>

Domain sequences are not given in the raw SCOP or CATH parsable files, but are
required for many analyses and for convenience should, ideally, be provided
along with the classification itself.  DOMAINSEQS reads a DCF file (domain 
classificaiton file) that lacks sequence information and writes one 
containing sequence information.





<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>

In order to find the start and end of a domain in a sequence from swissprot, 
the domain sequence from the domain CCF file is aligned to the 
full length protein sequence from swissprot.  Alignment is performed first
by string handling and if that fails, by using the EMBOSS implementation of
the Needleman and Wunsch global alignment algorithm.  Gap insertion and 
extension penalties used in the alignments are user-specified.




<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
     (Take this from "Sister applications" in the old documentation)
-->

<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<!--#include file="inc/domainseqs.isee" -->


<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
None.






<!-- AUTHORS -->
<a name="13.0"><br><br><br><H2> 13.0  AUTHORS                   </H2></a>
 Ranjeeva Ranasinghe  <a href="mailto:rranasin@hgmp.mrc.ac.uk">(rranasin@hgmp.mrc.ac.uk)</a>
<br><br>Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>MRC Rosalind Franklin Centre for Genomics Research, Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK





<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

<!--#include file="inc/embossref.ihtml" -->

<H3>14.1 Other useful references  </H3>

<br>

</BODY>
</HTML>
