<HTML>
<HEAD>
  <TITLE>
  EMBASSY: IPRSCAn: eiprscan
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
eiprscan
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
Motif detection
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>




<H2>
    Algorithm
</H2>





<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with eiprscan</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>eiprscan -nocrc -goterms -iprlookup </b>
Motif detection
Input sequence set: <b>iprtest.seq</b>
Applications to use
          all : all
  blastprodom : blastprodom
        coils : coils
       gene3d : gene3d
   hmmpanther : hmmpanther
       hmmpir : hmmpir
      hmmpfam : hmmpfam
     hmmsmart : hmmsmart
      hmmtigr : hmmtigr
   fprintscan : fprintscan
   scanregexp : scanregexp
  profilescan : profilescan
  superfamily : superfamily
          seg : seg
      signalp : signalp
        tmhmm : tmhmm
Application(s) to use [all]: <b></b>
Output Formats
       raw : Raw
       txt : Text
      html : Html
       xml : Xml
    ebixml : EBIxml
       gff : GFF
Format to use [xml]: <b>raw</b>
EIPRSCAN program output file [iprtest.eiprscan]: <b></b>
SUBMITTED-iprscan-20100115-12345678

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
Motif detection
Version: EMBOSS:6.2.0

   Standard (Mandatory) qualifiers:
  [-sequence]          seqset     Sequence set filename and optional format,
                                  or reference (input USA)
   -appl               menu       [all] Application(s) to use (Values: all
                                  (all); blastprodom (blastprodom); coils
                                  (coils); gene3d (gene3d); hmmpanther
                                  (hmmpanther); hmmpir (hmmpir); hmmpfam
                                  (hmmpfam); hmmsmart (hmmsmart); hmmtigr
                                  (hmmtigr); fprintscan (fprintscan);
                                  scanregexp (scanregexp); profilescan
                                  (profilescan); superfamily (superfamily);
                                  seg (seg); signalp (signalp); tmhmm (tmhmm))
   -format             menu       [xml] Format to use (Values: raw (Raw); txt
                                  (Text); html (Html); xml (Xml); ebixml
                                  (EBIxml); gff (GFF))
  [-outfile]           outfile    [*.eiprscan] EIPRSCAN program output file

   Additional (Optional) qualifiers (* if not always prompted):
   -email              string     Submitter email address (Any string)
   -trtable            menu       [0] Genetic codes used for translation
                                  (Values: 0 (Standard); 1 (Standard (with
                                  alternative initiation codons)); 2
                                  (Vertebrate Mitochondrial); 3 (Yeast
                                  Mitochondrial); 4 (Mold, Protozoan,
                                  Coelenterate Mitochondrial and
                                  Mycoplasma/Spiroplasma); 5 (Invertebrate
                                  Mitochondrial); 6 (Ciliate Macronuclear and
                                  Dasycladacean); 9 (Echinoderm
                                  Mitochondrial); 10 (Euplotid Nuclear); 11
                                  (Bacterial); 12 (Alternative Yeast Nuclear);
                                  13 (Ascidian Mitochondrial); 14 (Flatworm
                                  Mitochondrial); 15 (Blepharisma
                                  Macronuclear); 16 (Chlorophycean
                                  Mitochondrial); 21 (Trematode
                                  Mitochondrial); 22 (Scenedesmus obliquus);
                                  23 (Thraustochytrium Mitochondrial))
   -trlen              integer    [1] Minimum size of Open Reading Frames
                                  (ORFs) in the translations. (Integer 1 or
                                  more)
   -iprlookup          boolean    [N] Turn on InterPro lookup for results
*  -goterms            boolean    [N] Show GO terms in InterPro lookup
   -[no]crc            boolean    [Y] Perform CRC64 check
   -altjobs            boolean    [N] Launch jobs alternately (chunk after
                                  chunk)

   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-sequence" associated qualifiers
   -sbegin1            integer    Start of each sequence to be used
   -send1              integer    End of each sequence to be used
   -sreverse1          boolean    Reverse (if DNA)
   -sask1              boolean    Ask for begin/end/reverse
   -snucleotide1       boolean    Sequence is nucleotide
   -sprotein1          boolean    Sequence is protein
   -slower1            boolean    Make lower case
   -supper1            boolean    Make upper case
   -sformat1           string     Input sequence format
   -sdbname1           string     Database name
   -sid1               string     Entryname
   -ufo1               string     UFO features
   -fformat1           string     Features format
   -fopenfile1         string     Features file name

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-sequence]<br>(Parameter 1)</td>
<td>seqset</td>
<td>Sequence set filename and optional format, or reference (input USA)</td>
<td>Readable set of sequences</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-appl</td>
<td>list</td>
<td>Application(s) to use</td>
<td><table><tr><td>all</td> <td><i>(all)</i></td></tr><tr><td>blastprodom</td> <td><i>(blastprodom)</i></td></tr><tr><td>coils</td> <td><i>(coils)</i></td></tr><tr><td>gene3d</td> <td><i>(gene3d)</i></td></tr><tr><td>hmmpanther</td> <td><i>(hmmpanther)</i></td></tr><tr><td>hmmpir</td> <td><i>(hmmpir)</i></td></tr><tr><td>hmmpfam</td> <td><i>(hmmpfam)</i></td></tr><tr><td>hmmsmart</td> <td><i>(hmmsmart)</i></td></tr><tr><td>hmmtigr</td> <td><i>(hmmtigr)</i></td></tr><tr><td>fprintscan</td> <td><i>(fprintscan)</i></td></tr><tr><td>scanregexp</td> <td><i>(scanregexp)</i></td></tr><tr><td>profilescan</td> <td><i>(profilescan)</i></td></tr><tr><td>superfamily</td> <td><i>(superfamily)</i></td></tr><tr><td>seg</td> <td><i>(seg)</i></td></tr><tr><td>signalp</td> <td><i>(signalp)</i></td></tr><tr><td>tmhmm</td> <td><i>(tmhmm)</i></td></tr></table></td>
<td>all</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-format</td>
<td>list</td>
<td>Format to use</td>
<td><table><tr><td>raw</td> <td><i>(Raw)</i></td></tr><tr><td>txt</td> <td><i>(Text)</i></td></tr><tr><td>html</td> <td><i>(Html)</i></td></tr><tr><td>xml</td> <td><i>(Xml)</i></td></tr><tr><td>ebixml</td> <td><i>(EBIxml)</i></td></tr><tr><td>gff</td> <td><i>(GFF)</i></td></tr></table></td>
<td>xml</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outfile]<br>(Parameter 2)</td>
<td>outfile</td>
<td>EIPRSCAN program output file</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.eiprscan</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-email</td>
<td>string</td>
<td>Submitter email address</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-trtable</td>
<td>list</td>
<td>Genetic codes used for translation</td>
<td><table><tr><td>0</td> <td><i>(Standard)</i></td></tr><tr><td>1</td> <td><i>(Standard (with alternative initiation codons))</i></td></tr><tr><td>2</td> <td><i>(Vertebrate Mitochondrial)</i></td></tr><tr><td>3</td> <td><i>(Yeast Mitochondrial)</i></td></tr><tr><td>4</td> <td><i>(Mold, Protozoan, Coelenterate Mitochondrial and Mycoplasma/Spiroplasma)</i></td></tr><tr><td>5</td> <td><i>(Invertebrate Mitochondrial)</i></td></tr><tr><td>6</td> <td><i>(Ciliate Macronuclear and Dasycladacean)</i></td></tr><tr><td>9</td> <td><i>(Echinoderm Mitochondrial)</i></td></tr><tr><td>10</td> <td><i>(Euplotid Nuclear)</i></td></tr><tr><td>11</td> <td><i>(Bacterial)</i></td></tr><tr><td>12</td> <td><i>(Alternative Yeast Nuclear)</i></td></tr><tr><td>13</td> <td><i>(Ascidian Mitochondrial)</i></td></tr><tr><td>14</td> <td><i>(Flatworm Mitochondrial)</i></td></tr><tr><td>15</td> <td><i>(Blepharisma Macronuclear)</i></td></tr><tr><td>16</td> <td><i>(Chlorophycean Mitochondrial)</i></td></tr><tr><td>21</td> <td><i>(Trematode Mitochondrial)</i></td></tr><tr><td>22</td> <td><i>(Scenedesmus obliquus)</i></td></tr><tr><td>23</td> <td><i>(Thraustochytrium Mitochondrial)</i></td></tr></table></td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-trlen</td>
<td>integer</td>
<td>Minimum size of Open Reading Frames (ORFs) in the translations.</td>
<td>Integer 1 or more</td>
<td>1</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-iprlookup</td>
<td>boolean</td>
<td>Turn on InterPro lookup for results</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-goterms</td>
<td>boolean</td>
<td>Show GO terms in InterPro lookup</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]crc</td>
<td>boolean</td>
<td>Perform CRC64 check</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-altjobs</td>
<td>boolean</td>
<td>Launch jobs alternately (chunk after chunk)</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-sequence" associated seqset qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sbegin1<br>-sbegin_sequence</td>
<td>integer</td>
<td>Start of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -send1<br>-send_sequence</td>
<td>integer</td>
<td>End of each sequence to be used</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sreverse1<br>-sreverse_sequence</td>
<td>boolean</td>
<td>Reverse (if DNA)</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sask1<br>-sask_sequence</td>
<td>boolean</td>
<td>Ask for begin/end/reverse</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -snucleotide1<br>-snucleotide_sequence</td>
<td>boolean</td>
<td>Sequence is nucleotide</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sprotein1<br>-sprotein_sequence</td>
<td>boolean</td>
<td>Sequence is protein</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -slower1<br>-slower_sequence</td>
<td>boolean</td>
<td>Make lower case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -supper1<br>-supper_sequence</td>
<td>boolean</td>
<td>Make upper case</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sformat1<br>-sformat_sequence</td>
<td>string</td>
<td>Input sequence format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sdbname1<br>-sdbname_sequence</td>
<td>string</td>
<td>Database name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -sid1<br>-sid_sequence</td>
<td>string</td>
<td>Entryname</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ufo1<br>-ufo_sequence</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fformat1<br>-fformat_sequence</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fopenfile1<br>-fopenfile_sequence</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-outfile" associated outfile qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_outfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>


<b>eiprscan</b> reads any normal sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: iprtest.seq</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
&gt;RS16_ECOLI
MVTIRLARHGAKKRPFYQVVVADSRNARNGRFIERVGFFNPIASEKEEGTRLDLDRIAHW
VGQGATISDRVAALIKEVNKAA
&gt;Q9RHD9
XPKLEEGVEGLVHVSEMDWTNKNIHPSKVVQVGDEVEVQVLDIDEERRRISLGIKQCKSN
PWEDFSSQFNKGDRISGSIKSITDFGIFIGLDGGIDGLVHLSDISWNEVGEEAVRRFKKG
DELETVILSVDPERERISLGIKQLEDDPFSNYASLHEKGSIVRGTVKEVDAKGAVISLGD
DIEGILKASEISRDRVEDARNVLKEGEEVEAKIISIDRKSRVISLSVKSKDVDDEKDAMK
ELRKQEVESAGPTTIGDLIRAQMENQG
&gt;Y902_MYCTU Q10560 PROBABLE SENSOR-LIKE HISTIDINE KINASE RV0902C (EC 2.7.3.-).
MNILSRIFARTPSLRTRVVVATAIGAAIPVLIVGTVVWVGITNDRKERLDRRLDEAAGFA
IPFVPRGLDEIPRSPNDQDALITVRRGNVIKSNSDITLPKLQDDYADTYVRGVRYRVRTV
EIPGPEPTSVAVGATYDATVAETNNLHRRVLLICTFAIGAAAVFAWLLAAFAVRPFKQLA
EQTRSIDAGDEAPRVEVHGASEAIEIAEAMRGMLQRIWNEQNRTKEALASARDFAAVSSH
ELRTPLTAMRTNLEVLSTLDLPDDQRKEVLNDVIRTQSRIEATLSALERLAQGELSTSDD
HVPVDITDLLDRAAHDAARIYPDLDVSLVPSPTCIIVGLPAGLRLAVDNAIANAVKHGGA
TLVQLSAVSSRAGVEIAIDDNGSGVPEGERQVVFERFSRGSTASHSGSGLGLALVAQQAQ
LHGGTASLENSPLGGARLVLRLPGPS
</pre>
</td></tr></table><p>





<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:

        #include file="inc/reportformats.ihtml"
   -->

<b>eiprscan</b> 
outputs a file in a format selected from the commandline options


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: iprtest.eiprscan</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
RS16_ECOLI	F94D07049A6D489D	82	BlastProDom	PD003791	RS16_BUCAI_P57474;	9	77	1e-19	T	07-Jan-2010	IPR000307	Ribosomal protein S16	Molecular Function: structural constituent of ribosome (GO:0003735), Cellular Component: intracellular (GO:0005622), Cellular Component: ribosome (GO:0005840), Biological Process: translation (GO:0006412)
Q9RHD9	D44DAE8C544CB7C1	267	Coil	coil	coiled-coil	225	246	NA	?	07-Jan-2010	NULL	NULL
RS16_ECOLI	F94D07049A6D489D	82	Gene3D	G3DSA:3.30.1320.10	no description	1	77	1.7e-25	T	07-Jan-2010	IPR000307	Ribosomal protein S16	Molecular Function: structural constituent of ribosome (GO:0003735), Cellular Component: intracellular (GO:0005622), Cellular Component: ribosome (GO:0005840), Biological Process: translation (GO:0006412)
Q9RHD9	D44DAE8C544CB7C1	267	Gene3D	G3DSA:2.40.50.140	no description	3	57	4.5e-10	T	07-Jan-2010	IPR012340	Nucleic acid-binding, OB-fold	
Q9RHD9	D44DAE8C544CB7C1	267	Gene3D	G3DSA:2.40.50.140	no description	67	144	1.3e-16	T	07-Jan-2010	IPR012340	Nucleic acid-binding, OB-fold	
Q9RHD9	D44DAE8C544CB7C1	267	Gene3D	G3DSA:2.40.50.140	no description	156	230	9.2e-17	T	07-Jan-2010	IPR012340	Nucleic acid-binding, OB-fold	
Y902_MYCTU	CD84A335CCFFE6D7	446	Gene3D	G3DSA:3.30.565.10	no description	301	443	6.3e-25	T	07-Jan-2010	IPR003594	ATP-binding region, ATPase-like	Molecular Function: ATP binding (GO:0005524)
Y902_MYCTU	CD84A335CCFFE6D7	446	HMMPanther	PTHR23283:SF23	SENSORY TRANSDUCTION HISTIDINE KINASE (BACTERIAL SENSOR PROTEIN)	6	57	1.9e-48	T	07-Jan-2010	NULL	NULL	
Y902_MYCTU	CD84A335CCFFE6D7	446	HMMPanther	PTHR23283:SF23	SENSORY TRANSDUCTION HISTIDINE KINASE (BACTERIAL SENSOR PROTEIN)	151	443	1.9e-48	T	07-Jan-2010	NULL	NULL	
Y902_MYCTU	CD84A335CCFFE6D7	446	HMMPanther	PTHR23283	SENSOR HISTIDINE KINASE-RELATED	6	57	1.9e-48	T	07-Jan-2010	NULL	NULL	
Y902_MYCTU	CD84A335CCFFE6D7	446	HMMPanther	PTHR23283	SENSOR HISTIDINE KINASE-RELATED	151	443	1.9e-48	T	07-Jan-2010	NULL	NULL	
RS16_ECOLI	F94D07049A6D489D	82	HMMPanther	PTHR12919	30S RIBOSOMAL PROTEIN S16	1	78	4.2e-36	T	07-Jan-2010	IPR000307	Ribosomal protein S16	Molecular Function: structural constituent of ribosome (GO:0003735), Cellular Component: intracellular (GO:0005622), Cellular Component: ribosome (GO:0005840), Biological Process: translation (GO:0006412)
Q9RHD9	D44DAE8C544CB7C1	267	HMMPanther	PTHR23270	PROGRAMMED CELL DEATH PROTEIN 11 (PRE-RRNA PROCESSING PROTEIN RRP5)	72	239	2.6e-09	T	07-Jan-2010	NULL	NULL	
RS16_ECOLI	F94D07049A6D489D	82	HMMPfam	PF00886	Ribosomal_S16	8	68	1.5e-35	T	07-Jan-2010	IPR000307	Ribosomal protein S16	Molecular Function: structural constituent of ribosome (GO:0003735), Cellular Component: intracellular (GO:0005622), Cellular Component: ribosome (GO:0005840), Biological Process: translation (GO:0006412)
Q9RHD9	D44DAE8C544CB7C1	267	HMMPfam	PF00575	S1	1	55	1.9e-08	T	07-Jan-2010	IPR003029	S1, RNA binding	Molecular Function: RNA binding (GO:0003723)
Q9RHD9	D44DAE8C544CB7C1	267	HMMPfam	PF00575	S1	68	142	2.2e-21	T	07-Jan-2010	IPR003029	S1, RNA binding	Molecular Function: RNA binding (GO:0003723)
Q9RHD9	D44DAE8C544CB7C1	267	HMMPfam	PF00575	S1	155	228	9.7e-22	T	07-Jan-2010	IPR003029	S1, RNA binding	Molecular Function: RNA binding (GO:0003723)
Y902_MYCTU	CD84A335CCFFE6D7	446	HMMPfam	PF00672	HAMP	151	219	4.7e-11	T	07-Jan-2010	IPR003660	Signal transduction histidine kinase, HAMP region	Molecular Function: signal transducer activity (GO:0004871), Biological Process: signal transduction (GO:0007165), Cellular Component: membrane (GO:0016020)
Y902_MYCTU	CD84A335CCFFE6D7	446	HMMPfam	PF00512	HisKA	230	296	1.3e-13	T	07-Jan-2010	IPR003661	Signal transduction histidine kinase, subgroup 1, dimerisation and phosphoacceptor region	Molecular Function: two-component sensor activity (GO:0000155), Biological Process: signal transduction (GO:0007165), Cellular Component: membrane (GO:0016020)
Y902_MYCTU	CD84A335CCFFE6D7	446	HMMPfam	PF02518	HATPase_c	338	445	3.4e-29	T	07-Jan-2010	IPR003594	ATP-binding region, ATPase-like	Molecular Function: ATP binding (GO:0005524)
Q9RHD9	D44DAE8C544CB7C1	267	HMMSmart	SM00316	no description	3	55	1.2e-06	T	07-Jan-2010	IPR003029	S1, RNA binding	Molecular Function: RNA binding (GO:0003723)
Q9RHD9	D44DAE8C544CB7C1	267	HMMSmart	SM00316	no description	70	142	1.4e-19	T	07-Jan-2010	IPR003029	S1, RNA binding	Molecular Function: RNA binding (GO:0003723)
Q9RHD9	D44DAE8C544CB7C1	267	HMMSmart	SM00316	no description	157	228	2.6e-21	T	07-Jan-2010	IPR003029	S1, RNA binding	Molecular Function: RNA binding (GO:0003723)
Y902_MYCTU	CD84A335CCFFE6D7	446	HMMSmart	SM00304	no description	170	222	3.1e-06	T	07-Jan-2010	IPR003660	Signal transduction histidine kinase, HAMP region	Molecular Function: signal transducer activity (GO:0004871), Biological Process: signal transduction (GO:0007165), Cellular Component: membrane (GO:0016020)
Y902_MYCTU	CD84A335CCFFE6D7	446	HMMSmart	SM00388	no description	230	296	2.4e-12	T	07-Jan-2010	IPR003661	Signal transduction histidine kinase, subgroup 1, dimerisation and phosphoacceptor region	Molecular Function: two-component sensor activity (GO:0000155), Biological Process: signal transduction (GO:0007165), Cellular Component: membrane (GO:0016020)
Y902_MYCTU	CD84A335CCFFE6D7	446	HMMSmart	SM00387	no description	338	446	5e-24	T	07-Jan-2010	IPR003594	ATP-binding region, ATPase-like	Molecular Function: ATP binding (GO:0005524)
RS16_ECOLI	F94D07049A6D489D	82	HMMTigr	TIGR00002	S16: ribosomal protein S16	2	81	1.8e-32	T	07-Jan-2010	IPR000307	Ribosomal protein S16	Molecular Function: structural constituent of ribosome (GO:0003735), Cellular Component: intracellular (GO:0005622), Cellular Component: ribosome (GO:0005840), Biological Process: translation (GO:0006412)
Q9RHD9	D44DAE8C544CB7C1	267	FPrintScan	PR00681	RIBOSOMALS1	6	27	7.9e-17	T	07-Jan-2010	IPR000110	Ribosomal protein S1	Molecular Function: RNA binding (GO:0003723), Molecular Function: structural constituent of ribosome (GO:0003735), Cellular Component: ribosome (GO:0005840), Biological Process: translation (GO:0006412)
Q9RHD9	D44DAE8C544CB7C1	267	FPrintScan	PR00681	RIBOSOMALS1	85	104	7.9e-17	T	07-Jan-2010	IPR000110	Ribosomal protein S1	Molecular Function: RNA binding (GO:0003723), Molecular Function: structural constituent of ribosome (GO:0003735), Cellular Component: ribosome (GO:0005840), Biological Process: translation (GO:0006412)
Q9RHD9	D44DAE8C544CB7C1	267	FPrintScan	PR00681	RIBOSOMALS1	125	143	7.9e-17	T	07-Jan-2010	IPR000110	Ribosomal protein S1	Molecular Function: RNA binding (GO:0003723), Molecular Function: structural constituent of ribosome (GO:0003735), Cellular Component: ribosome (GO:0005840), Biological Process: translation (GO:0006412)
Y902_MYCTU	CD84A335CCFFE6D7	446	FPrintScan	PR00344	BCTRLSENSOR	374	388	1.1e-11	T	07-Jan-2010	IPR004358	Signal transduction histidine kinase-related protein, C-terminal	Biological Process: phosphorylation (GO:0016310), Molecular Function: transferase activity, transferring phosphorus-containing groups (GO:0016772)
Y902_MYCTU	CD84A335CCFFE6D7	446	FPrintScan	PR00344	BCTRLSENSOR	392	402	1.1e-11	T	07-Jan-2010	IPR004358	Signal transduction histidine kinase-related protein, C-terminal	Biological Process: phosphorylation (GO:0016310), Molecular Function: transferase activity, transferring phosphorus-containing groups (GO:0016772)
Y902_MYCTU	CD84A335CCFFE6D7	446	FPrintScan	PR00344	BCTRLSENSOR	406	424	1.1e-11	T	07-Jan-2010	IPR004358	Signal transduction histidine kinase-related protein, C-terminal	Biological Process: phosphorylation (GO:0016310), Molecular Function: transferase activity, transferring phosphorus-containing groups (GO:0016772)
Y902_MYCTU	CD84A335CCFFE6D7	446	FPrintScan	PR00344	BCTRLSENSOR	430	443	1.1e-11	T	07-Jan-2010	IPR004358	Signal transduction histidine kinase-related protein, C-terminal	Biological Process: phosphorylation (GO:0016310), Molecular Function: transferase activity, transferring phosphorus-containing groups (GO:0016772)
RS16_ECOLI	F94D07049A6D489D	82	ScanRegExp	PS00732	RIBOSOMAL_S16	2	11	8e-5	T	07-Jan-2010	IPR000307	Ribosomal protein S16	Molecular Function: structural constituent of ribosome (GO:0003735), Cellular Component: intracellular (GO:0005622), Cellular Component: ribosome (GO:0005840), Biological Process: translation (GO:0006412)
Q9RHD9	D44DAE8C544CB7C1	267	ProfileScan	PS50126	S1	1	55	14.869	T	07-Jan-2010	IPR003029	S1, RNA binding	Molecular Function: RNA binding (GO:0003723)
Q9RHD9	D44DAE8C544CB7C1	267	ProfileScan	PS50126	S1	72	142	20.809	T	07-Jan-2010	IPR003029	S1, RNA binding	Molecular Function: RNA binding (GO:0003723)
Q9RHD9	D44DAE8C544CB7C1	267	ProfileScan	PS50126	S1	159	228	22.541	T	07-Jan-2010	IPR003029	S1, RNA binding	Molecular Function: RNA binding (GO:0003723)
Y902_MYCTU	CD84A335CCFFE6D7	446	ProfileScan	PS50885	HAMP	170	222	7.777	T	07-Jan-2010	IPR003660	Signal transduction histidine kinase, HAMP region	Molecular Function: signal transducer activity (GO:0004871), Biological Process: signal transduction (GO:0007165), Cellular Component: membrane (GO:0016020)
Y902_MYCTU	CD84A335CCFFE6D7	446	ProfileScan	PS50109	HIS_KIN	237	446	34.449	T	07-Jan-2010	IPR005467	Signal transduction histidine kinase, core	Molecular Function: two-component sensor activity (GO:0000155), Molecular Function: protein histidine kinase activity (GO:0004673), Biological Process: signal transduction (GO:0007165), Biological Process: peptidyl-histidine phosphorylation (GO:0018106)
RS16_ECOLI	F94D07049A6D489D	82	superfamily	SSF54565	Ribosomal protein S16	1	79	3.1e-27	T	07-Jan-2010	IPR000307	Ribosomal protein S16	Molecular Function: structural constituent of ribosome (GO:0003735), Cellular Component: intracellular (GO:0005622), Cellular Component: ribosome (GO:0005840), Biological Process: translation (GO:0006412)
Q9RHD9	D44DAE8C544CB7C1	267	superfamily	SSF50249	Nucleic acid-binding proteins	67	220	7.6e-41	T	07-Jan-2010	IPR016027	Nucleic acid-binding, OB-fold-like	
Q9RHD9	D44DAE8C544CB7C1	267	superfamily	SSF50249	Nucleic acid-binding proteins	3	61	1.4e-14	T	07-Jan-2010	IPR016027	Nucleic acid-binding, OB-fold-like	
Y902_MYCTU	CD84A335CCFFE6D7	446	superfamily	SSF55874	ATPase domain of HSP90 chaperone/DNA topoisomerase II/histidine kinase	303	444	8e-31	T	07-Jan-2010	IPR003594	ATP-binding region, ATPase-like	Molecular Function: ATP binding (GO:0005524)
Y902_MYCTU	CD84A335CCFFE6D7	446	superfamily	SSF47384	Homodimeric domain of signal transducing histidine kinase	220	292	4.8e-12	T	07-Jan-2010	IPR009082	Signal transduction histidine kinase, homodimeric	Molecular Function: signal transducer activity (GO:0004871), Biological Process: signal transduction (GO:0007165)
Q9RHD9	D44DAE8C544CB7C1	267	Seg	seg	seg	29	40	NA	?	07-Jan-2010	NULL	NULL
Q9RHD9	D44DAE8C544CB7C1	267	Seg	seg	seg	84	98	NA	?	07-Jan-2010	NULL	NULL
Q9RHD9	D44DAE8C544CB7C1	267	Seg	seg	seg	222	237	NA	?	07-Jan-2010	NULL	NULL
Y902_MYCTU	CD84A335CCFFE6D7	446	Seg	seg	seg	44	55	NA	?	07-Jan-2010	NULL	NULL
Y902_MYCTU	CD84A335CCFFE6D7	446	Seg	seg	seg	108	120	NA	?	07-Jan-2010	NULL	NULL
Y902_MYCTU	CD84A335CCFFE6D7	446	Seg	seg	seg	160	173	NA	?	07-Jan-2010	NULL	NULL
Y902_MYCTU	CD84A335CCFFE6D7	446	Seg	seg	seg	308	319	NA	?	07-Jan-2010	NULL	NULL
Y902_MYCTU	CD84A335CCFFE6D7	446	Seg	seg	seg	400	424	NA	?	07-Jan-2010	NULL	NULL
</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

None other rthan the databases configured for the
native <i>iprscan</i> application.






<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="/emboss/apps/antigenic.html">antigenic</a></td>
<td>Finds antigenic sites in proteins</td>
</tr>

<tr>
<td><a href="/emboss/apps/digest.html">digest</a></td>
<td>Reports on protein proteolytic enzyme or reagent cleavage sites</td>
</tr>

<tr>
<td><a href="/embassy/cbstools/echlorop.html">echlorop</a></td>
<td>Reports presence of chloroplast transit peptides</td>
</tr>

<tr>
<td><a href="/embassy/cbstools/elipop.html">elipop</a></td>
<td>Prediction of lipoproteins</td>
</tr>

<tr>
<td><a href="/embassy/memenew/emast.html">emast</a></td>
<td>Motif detection</td>
</tr>

<tr>
<td><a href="/embassy/memenew/ememe.html">ememe</a></td>
<td>Multiple EM for Motif Elicitation</td>
</tr>

<tr>
<td><a href="/embassy/memenew/ememetext.html">ememetext</a></td>
<td>Multiple EM for Motif Elicitation. Text file only</td>
</tr>

<tr>
<td><a href="/embassy/cbstools/enetnglyc.html">enetnglyc</a></td>
<td>Reports N-glycosylation sites in human proteins</td>
</tr>

<tr>
<td><a href="/embassy/cbstools/enetoglyc.html">enetoglyc</a></td>
<td>Reports mucin type GalNAc O-glycosylation sites in mammalian proteins</td>
</tr>

<tr>
<td><a href="/embassy/cbstools/enetphos.html">enetphos</a></td>
<td>Reports ser, thr and tyr phosphorylation sites in eukaryotic proteins</td>
</tr>

<tr>
<td><a href="/emboss/apps/epestfind.html">epestfind</a></td>
<td>Finds PEST motifs as potential proteolytic cleavage sites</td>
</tr>

<tr>
<td><a href="/embassy/cbstools/eprop.html">eprop</a></td>
<td>Reports propeptide cleavage sites in proteins</td>
</tr>

<tr>
<td><a href="/embassy/cbstools/esignalp.html">esignalp</a></td>
<td>Reports protein signal cleavage sites</td>
</tr>

<tr>
<td><a href="/embassy/cbstools/etmhmm.html">etmhmm</a></td>
<td>Reports transmembrane helices</td>
</tr>

<tr>
<td><a href="/embassy/cbstools/eyinoyang.html">eyinoyang</a></td>
<td>Reports O-(beta)-GlcNAc attachment sites</td>
</tr>

<tr>
<td><a href="/emboss/apps/fuzzpro.html">fuzzpro</a></td>
<td>Search for patterns in protein sequences</td>
</tr>

<tr>
<td><a href="/emboss/apps/fuzztran.html">fuzztran</a></td>
<td>Search for patterns in protein sequences (translated)</td>
</tr>

<tr>
<td><a href="/emboss/apps/helixturnhelix.html">helixturnhelix</a></td>
<td>Identify nucleic acid-binding motifs in protein sequences</td>
</tr>

<tr>
<td><a href="/emboss/apps/oddcomp.html">oddcomp</a></td>
<td>Identify proteins with specified sequence word composition</td>
</tr>

<tr>
<td><a href="/embassy/meme/omeme.html">omeme</a></td>
<td>Motif detection</td>
</tr>

<tr>
<td><a href="/emboss/apps/patmatdb.html">patmatdb</a></td>
<td>Searches protein sequences with a sequence motif</td>
</tr>

<tr>
<td><a href="/emboss/apps/patmatmotifs.html">patmatmotifs</a></td>
<td>Scan a protein sequence with motifs from the PROSITE database</td>
</tr>

<tr>
<td><a href="/emboss/apps/pepcoil.html">pepcoil</a></td>
<td>Predicts coiled coil regions in protein sequences</td>
</tr>

<tr>
<td><a href="/emboss/apps/preg.html">preg</a></td>
<td>Regular expression search of protein sequence(s)</td>
</tr>

<tr>
<td><a href="/emboss/apps/pscan.html">pscan</a></td>
<td>Scans protein sequence(s) with fingerprints from the PRINTS database</td>
</tr>

<tr>
<td><a href="/emboss/apps/sigcleave.html">sigcleave</a></td>
<td>Reports on signal cleavage sites in a protein sequence</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

This program is an EMBOSS wrapper for the InterProSCAN application
"iprscan"

<p>
The original iprscan application must be installed and configured to
use this wrapper.


<p>
Although we take every care to ensure that the results of the EMBOSS
version are identical to those from the original package, we recommend
that you check your inputs give the same results in both versions
before publication.

<p>
Please report all bugs in the EMBOSS version to the EMBOSS bug team,
not to the original author.

<H2>
    History
</H2>





<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>

None

</BODY>
</HTML>

