<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: INTERFACE documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<!--#include virtual="/apps/header1.inc" -->
<H2> INTERFACE documentation
<!--#include virtual="/apps/header2.inc" -->

<!-- END OF HEADER -->




<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>


<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>
Reads CCF files (clean coordinate files) and writes CON files (contact files) of inter-chain residue-residue contact data.

<!--#include file="inc/interface.ione" -->



<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
INTERFACE reads a protein CCF file (clean coordinate file) and writes a CON file (contacts file) of inter-chain residue-residue contact data.  The file contains residue contact data for all pairs of chains in each model in the CCF file.  The input and output files are specified by the user.  A log file is also written.


<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.
     Use "<b>INTERFACE</b> reads any normal sequence USAs." if
     appropriate.
-->



<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>
The format of the protein CCF file is described in the <a href="pdbparse.html">PDBPARSE documentation.</a>

<!--#include file="inc/interface.input" -->






<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>

The CON format used for the contact files (Figure 1) is similar to EMBL format and is described in the <a href="contacts.html">CONTACTS documentation.</a>  A few of the records differ in the INTERFACE output compared to the CONTACTS output, however, so for the sake of clarity all records are described below.

<UL>
<LI> XX - used for spacing and comments. The first line is bibliographic information and contains
 the text "Inter-chain residue-residue contact data".
<LI> TY - type of contact.  For CON files generated by CONTACTS, 'INTER' is always given.
<LI>  EX - experimental information. The value of the threshold contact 
distance is given as a floating point number after 'THRESH'. For CON files generated by 
INTERFACE, a '.' is given after 'THRESH' (this record is used by the CONTACTS application and
can be disregarded). The number of models and number of polypeptide chains are given after 
'NMOD' and 'NCHA' respectively. For domain CCF files a 1 is always given.
<LI> NE - number of entries (interfaces) in the file. For CON files generated by INTERFACE this is 
the number of interfaces (pairs of chains), equal to the number of models multiplied by
the number of unique interfaces.
Following the NE record, the file has a section for each entry containing
records for entry number (EN), identifier codes (ID),  
polypeptide chain-specific data (CN), chain sequence information (S1 & S2) and 
number of contacts (NC), that together define the interface and its contacts. 
<LI> EN - entry number.  The number in brackets indicates the start of an entry (interface).
<LI> CN - polypeptide chain-specific data.  Tokens delimiting data items are as follow. 
 (1) MO: The model number (from the PDB file).  (2) CN1: Chain number (first chain). 
(3) CN2: Chain number (second chain).  (4) ID1: PDB chain identifier (first chain)
(a '.` given in cases where a chain identifier was not specified in the original PDB file or, for domain CCF files, the domain 
is comprised of more than one domain). (5) ID2: Chain identifier (second chain). (6) NRES1: number of residues (first chain). 
(7) NRES2: number of residues (second chain).
<LI> NC - number of contacts: (1) SM: Number of residue-residue contacts; between side-chain or 
main-chain atoms of an amino acid residue in one chain and the side- or main-chain atoms of a 
residue in a different chain. (2) LI: Number of residue-ligand contacts; between side-chain or 
main-chain atoms of an amino acid residue and a ligand.
<LI> ID - identifier codes: (1) PDB: 4-character PDB identifier code. (2) DOM: SCOP or CATH domain identifier
code. (3) LIG: Not used by INTERFACE, a '.' is given. 
<LI> S1 - polypeptide chain sequence (first chain). The number of residues is given 
before AA on the first line.  The sequece is given on subsequent lines. 
<LI> S2 - polypeptide chain sequence (second chain).
<LI> SM - Line of residue contact data. Pairs of amino acid identifiers and 
residue numbers are delimited by a ';'. The left-hand term refers to the first chain and 
the right-hand term the second chain.  Residue numbers are taken from the 
CCF files and give a correct index into the sequences (i.e. they
are not necessarily the same as the original PDB file).  These sequence is 
given in the CON file itself (S1 & S2 records).
<LI> // - delimiter for individual entries in the file and also given on
the last line of the file.
</UL> 


<!--#include file="inc/interface.output" -->



<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>

INTERFACE uses a data file containing van der Waals radii for atoms in 
proteins (see <a href="contacts.html">CONTACTS documentation.</a>)
The file Evdw.dat is such a data file and is part of the EMBOSS distribution.
<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" >
<tr><td>
<pre>
<!--#include file="inc/interface.ihelp" --></pre>
</pre>
</td></tr>
</table>


<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 
<!--#include file="inc/interface.itable" -->

<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of INTERFACE is shown below.

<!--#include file="inc/interface.usage" -->

<!-- <br><br>
The protein CCF file /test_data/interface/2hhb.ccf was read and the CON file /test_data/interface/2hhb.con was written.  van der Waals radii were taken from the emboss data file Evdw.dat which is specified as a default value in interface.acd and the default contact distance of 1 Angstrom was used to determine contacts.  The log file test_data/interface/interface.log was written.
<br><br>


<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
None.








<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>

<b> Types of contact </b>
<br><b>SM</b> records are used for contacts between either either side-chain or main-chain atoms 
as defined above.  In a future implementation, <b>SS</b> will be used for side-chain only contacts,
 <b>MM</b> will be used for main-chain only contacts, <b>LI</b> will be used for contacts to
ligands, and there will probably be several other forms of contact too.

<br><br><b>Threshold ignore distance option</b>
<br>The threshold ignore distance can be adjusted to speed-up the calculation of contacts considerably.  If any two atoms from two different residues are at least this distance apart then no futher inter-atomic contacts will be checked for for that residue pair. 


<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b> Clean coordinate file (for protein) </b></td>
<td> CCF format (EMBL-like).</td>
<td> Protein coordinate and derived data for a single PDB file.  The data are 'cleaned-up': self-consistent and error-corrected.  </td>
<td> <a href="pdbparse.html">PDBPARSE</a> </td>
<td> Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS. </td>
</tr>
<tr>
<td><b> Clean coordinate file (for domain)</b></td>
<td> CCF format (EMBL-like).</td>
<td> Protein coordinate and derived data for a single domain from SCOP or CATH. The data are 'cleaned-up': self-consistent and error-corrected. </td>
<td> <a href="domainer.html">DOMAINER</a></td>
<td> Records for residue solvent accessibility and secondary structure are added to the file by using PDBPLUS.</td>
</tr>
<tr>
<td><b> Contact file (intra-chain residue-residue contacts)</b></td>
<td> CON format (EMBL-like.) </td>
<td> Intra-chain residue-residue contact data for a protein or a domain from SCOP or CATH.  </td>
<td> <a href="contacts.html">CONTACTS</a> </td>
<td> N.A. </td>
</tr>
<tr>
<td><b> Contact file (inter-chain residue-residue contacts)</b></td>
<td> CON format (EMBL-like.) </td>
<td> Inter-chain residue-residue contact data for a protein or a domain from SCOP or CATH.  </td>
<td> <a href="interface.html">INTERFACE</a> </td>
<td> N.A. </td>
</tr>
<tr>
<td><b> Contact file (residue-ligand contacts)</b></td>
<td> CON format (EMBL-like.) </td>
<td> Residue-ligand contact data for a protein or a domain from SCOP or CATH. </td>
<td> <a href="sites.html">SITES</a> </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>van der Waals radii  </b></td>
<td> A file of van der Waals radii for atoms in amino acid residues. Part of the emboss distribution.</td>
<td> N.A. </td>
<td> N.A. </td>
</tr>
</table>


<!--#include file="inc/interface.history" -->
<!--#include file="inc/interface.comment" -->

<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>
Knowledge of the physical contacts that amino acid residues in two different polypeptide chains of a protein make with one another is required for several different analyses.  INTERFACE calculates inter-chain residue-residue contact data from protein CCF files (clean coordinate files).


<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>
Contact between two residues is defined as when the van der Waals surface of any atom of the first residue comes within the threshold contact distance of the van der Waals surface of any atom of the second residue. The threshold contact distance is a user-defined distance with a default value of 1 Angstrom.



<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
-->

<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<!--#include file="inc/pdbparse.isee" -->



<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
INTERFACE generates a log file an excerpt of which is shown below. If there is a problem in processing a CCF file, three lines containing the record '//', the PDB identifier code and an error message respectively are written. The text <i>'WARN  file open error filename' </i>,  <i>'ERROR file read error filename' </i> or  <i>'ERROR file write error filename ' </i> will be reported when an error was encountered during a file open, read or write respectively.  Various other error messages may also be given (in case of difficulty email Jon Ison, jison@hgmp.mrc.ac.uk).



<!-- AUTHORS
     The main author first then all others. 
-->
<a name="13.0"></a>
<br><br><br><H2> 13.0  AUTHORS                   </H2>
Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK




<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

<!--#include file="inc/embossref.ihtml" -->


<H3>14.1 Other useful references  </H3>


</BODY>
</HTML>



