<HTML>
<HEAD>
  <TITLE>
  EMBASSY: VIENNA: vrnadistance
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
vrnadistance
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>



<H2>
Wiki
</H2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

<H2>
    Function
</H2>
RNA distances
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>


This is a port of the Vienna RNA package program RNAdistance.


<H2>
    Algorithm
</H2>

See the original documentation for the Vienna RNA package
http://www.tbi.univie.ac.at/~ivo/RNA/



<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with vrnadistance</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>vrnadistance -distance 'f,h' -doalignment </b>
RNA distances
Vienna RNA structures file: <b>rna4.fold</b>
Vienna RNAfold output file [rna4.vrnadistance]: <b></b>

</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<H2>
    Command line arguments
</H2>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-structuresfile]    infile     Vienna RNA structures file
  [-outfile]           outfile    [*.vrnadistance] Vienna RNAfold output file

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers:
   -distance           menu       [f] Distance types to calculate (Values: f
                                  (Full (tree)); h (HIT (tree)); w (Weighted
                                  coarse (tree)); c (Coarse (tree)); F (Full
                                  (string)); H (HIT (string)); W (Weighted
                                  coarse (string)); C (Coarse (string)); P
                                  (Base pair distance))
   -compare            menu       [p] Distance types to calculate (Values: p
                                  (Pairwise); m (Distance matrix between all
                                  structures); f (Compare all structures to
                                  the first one); c (Compare continuously))
   -schapiro           boolean    [N] Use cost matrix for comparing coarse
                                  structures
   -doalignment        boolean    [N] Alignment showing matching substructures

   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-structuresfile]<br>(Parameter 1)</td>
<td>Vienna RNA structures file</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>Vienna RNAfold output file</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.vrnadistance</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td colspan=4>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-distance</td>
<td>Distance types to calculate</td>
<td><table><tr><td>f</td> <td><i>(Full (tree))</i></td></tr><tr><td>h</td> <td><i>(HIT (tree))</i></td></tr><tr><td>w</td> <td><i>(Weighted coarse (tree))</i></td></tr><tr><td>c</td> <td><i>(Coarse (tree))</i></td></tr><tr><td>F</td> <td><i>(Full (string))</i></td></tr><tr><td>H</td> <td><i>(HIT (string))</i></td></tr><tr><td>W</td> <td><i>(Weighted coarse (string))</i></td></tr><tr><td>C</td> <td><i>(Coarse (string))</i></td></tr><tr><td>P</td> <td><i>(Base pair distance)</i></td></tr></table></td>
<td>f</td>
</tr>

<tr>
<td>-compare</td>
<td>Distance types to calculate</td>
<td><table><tr><td>p</td> <td><i>(Pairwise)</i></td></tr><tr><td>m</td> <td><i>(Distance matrix between all structures)</i></td></tr><tr><td>f</td> <td><i>(Compare all structures to the first one)</i></td></tr><tr><td>c</td> <td><i>(Compare continuously)</i></td></tr></table></td>
<td>p</td>
</tr>

<tr>
<td>-schapiro</td>
<td>Use cost matrix for comparing coarse structures</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-doalignment</td>
<td>Alignment showing matching substructures</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>

<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->

<b>vrnadistance</b> reads any normal sequence USAs.

<p>


<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: rna4.fold</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
((.(((((((.....))))))).))....((..((((.....)))).)).
.....((((..((((..........)))).)))).....(((....))).
</pre>
</td></tr></table><p>





<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
#include file="inc/reportformats.ihtml"
   -->

<b>vrnadistance</b> 
outputs a graph to the specified graphics device. 
outputs a report format file. The default format is ...


<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: rna4.vrnadistance</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
f: 26  
(_____(.(((__((((....._____))))_))).))....((..((((.....)))).)).
_.....(_(((..((((..........)))).)))_)_....__.__(((...._)))_.___
h: 32  
(____((U1)((U5_)P7)(U1)P2)(U4)((U2)((U5)P4)(U1)P2)(U1)R1)
((U5)((U2)((U10)P4)(U1)P4)(U5)_____((U4)P3)(U1)_______R1)

</pre>
</td></tr></table><p>





<H2>
    Data files
</H2>

For details of Vienna RNA file formats, see the original documentation
http://www.tbi.univie.ac.at/~ivo/RNA/


<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->

None.







<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

None.








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->

None.







<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="/emboss/apps/einverted.html">einverted</a></td>
<td>Finds inverted repeats in nucleotide sequences</td>
</tr>

<tr>
<td><a href="vrnaalifold.html">vrnaalifold</a></td>
<td>RNA alignment folding</td>
</tr>

<tr>
<td><a href="vrnaalifoldpf.html">vrnaalifoldpf</a></td>
<td>RNA alignment folding with partition</td>
</tr>

<tr>
<td><a href="vrnacofold.html">vrnacofold</a></td>
<td>RNA cofolding</td>
</tr>

<tr>
<td><a href="vrnacofoldconc.html">vrnacofoldconc</a></td>
<td>RNA cofolding with concentrations</td>
</tr>

<tr>
<td><a href="vrnacofoldpf.html">vrnacofoldpf</a></td>
<td>RNA cofolding with partitioning</td>
</tr>

<tr>
<td><a href="vrnaduplex.html">vrnaduplex</a></td>
<td>RNA duplex calculation</td>
</tr>

<tr>
<td><a href="vrnaeval.html">vrnaeval</a></td>
<td>RNA eval</td>
</tr>

<tr>
<td><a href="vrnaevalpair.html">vrnaevalpair</a></td>
<td>RNA eval with cofold</td>
</tr>

<tr>
<td><a href="vrnafold.html">vrnafold</a></td>
<td>Calculate secondary structures of RNAs</td>
</tr>

<tr>
<td><a href="vrnafoldpf.html">vrnafoldpf</a></td>
<td>Secondary structures of RNAs with partition</td>
</tr>

<tr>
<td><a href="vrnaheat.html">vrnaheat</a></td>
<td>RNA melting</td>
</tr>

<tr>
<td><a href="vrnainverse.html">vrnainverse</a></td>
<td>RNA sequences matching a structure</td>
</tr>

<tr>
<td><a href="vrnalfold.html">vrnalfold</a></td>
<td>Calculate locally stable secondary structures of RNAs</td>
</tr>

<tr>
<td><a href="vrnaplot.html">vrnaplot</a></td>
<td>Plot vrnafold output</td>
</tr>

<tr>
<td><a href="vrnasubopt.html">vrnasubopt</a></td>
<td>Calculate RNA suboptimals</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

This program is an EMBOSS conversion of a program written by Ivo Hofacker
as part of his VIENNA package.

<p>
Although we take every care to ensure that the results of the EMBOSS
version are identical to those from the original package, we recommend
that you check your inputs give the same results in both versions
before publication.

<p>
Please report all bugs in the EMBOSS version to the EMBOSS bug team,
not to the original author.

<H2>
    History
</H2>

Converted (October 2005) by Alan Bleasby



<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<H2>
    Comments
</H2>


</BODY>
</HTML>

