<!--START OF HEADER - DON'T ALTER -->

<HTML>
<HEAD>
  <TITLE>
  EMBOSS: emast
  </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">



<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
emast
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>


<!--END OF HEADER-->






<H2>
    Function
</H2>
Motif detection
<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->




<H2>
    Description
</H2>

<!-- 
	This is the main part of the document.
        This is what the user looks at to find out if this program will do what he wants. 
        Long description.
        Why it is useful.
        When it should/shouldn't be used.
  -->
EMBASSY MEME is a suite of application wrappers to the original meme v3.0.14 applications written by  Timothy Bailey. meme v3.0.14 must be installed on the same system as EMBOSS and the location of the meme executables must be defined in your path for EMBASSY MEME to work.

<p><b>Usage:</b>
<br>ememe [options] mfile outfile
<p>The outfile parameter is new to EMBASSY MEME. The output is always written to <outfile>. 

<p>  MAST: Motif Alignment and Search Tool
  
<p>  MAST is a tool for searching biological sequence databases for sequences
  that contain one or more of a group of known motifs. 
  
<p>  A motif is a sequence pattern that occurs repeatedly in a group of related
  protein or DNA sequences. Motifs are represented as position-dependent
  scoring matrices that describe the score of each possible letter at each
  position in the pattern. Individual motifs may not contain gaps. Patterns with
  variable-length gaps must be split into two or more separate motifs before
  being submitted as input to MAST. 
  
<p>  MAST takes as input a file containing the descriptions of one or more motifs
  and searches a sequence database that you select for sequences that match
  the motifs. The motif file can be the output of the MEME motif discovery tool 
  or any file in the appropriate format. 
  
<p>  MAST outputs three things: 
<ul>  
<li>    1. The names of the high-scoring sequences sorted by the strength of the
       combined match of the sequence to all of the motifs in the group. 
<li>    2. Motif diagrams showing the order and spacing of the motifs within each
       matching sequence. 
<li>    3. Detailed annotation of each matching sequence showing the sequence
       and the locations and strengths of matches to the motifs. 
</ul>
  
<p>  MAST works by calculating match scores for each sequence in the database
  compared with each of the motifs in the group of motifs you provide. For each
  sequence, the match scores are converted into various types of p-values and
  these are used to determine the overall match of the sequence to the group of
  motifs and the probable order and spacing of occurrences of the motifs in the
  sequence. 












<H2>
    Algorithm
</H2>

<!-- 
        Algorithms
	  Pseudocode/Vague description/Handwaving/web links to useful pages
  -->

Please read the file README distributed with the original MEME. 












<H2>
    Usage
</H2>

<!--  
	Example usage, as run from the command-line.
        Many examples illustrating different behaviours is good.
 -->

<b>Here is a sample session with emast</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>emast ex1.html ex1.out </b>
Motif detection
Print results for sequences with E-value [10]: <b></b>
Show motif matches with p-value < mt [0.0001]: <b></b>


</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>



<h3>  EXAMPLES:</h3>
<p><b>Please note</b> the examples below are unedited excerpts of the original MEME documentation.  Bear in mind the EMBASSY and original MEME options may differ in practice (see "1. Command-line arguments").
  
<p>  The following examples assume that file "meme.results" is the
  output of a MEME run containing at least 3 motifs and file
  SwissProt is a copy of the Swiss-Prot database on your local disk.
  DNA_DB is a copy of a DNA database on your local disk.
   
<p>  1) Annotate the training set: 
<br>  	mast meme.results
   
<p>  2) Find sequences matching the motif and annotate them in the SwissProt database:</b>
<br>  	mast meme.results -d SwissProt
   
<p>  3) Show sequences with weaker combined matches to motifs.</b>
<br>  	mast meme.results -d SwissProt -ev 200
   
<p>  4) Indicate weaker matches to single motifs in the annotation so that sequences with weak matches to the motifs (but perhaps with the "correct" order and spacing) can be seen:
<br>  	mast meme.results -d SwissProt -w
   
<p>  5) Include a nominal order and spacing of the first three motifs in the calculation of the sequence p-values to increase the sensitivity of the search for matching sequences:
<br>  	mast meme.results -d SwissProt -diag "9-[2]-61-[1]-62-[3]-91"
   
<p>  6) Use only the first and third motifs in the search:
<br>  	mast meme.results -d SwissProt -m 1 -m 3
   
<p>  7) Use only the first two motifs in the search:
<br>  	mast meme.results -d SwissProt -c 2
  
<p>  8) Search DNA sequences using protein motifs, adjusting p-values and E-values for each sequence by that sequence's composition:
<br> 	mast meme.results -d DNA_DB -dna -comp



<H2>
    Command line arguments
</H2>
Where possible, the same command-line qualifier names and parameter order is used as in the original mast. There are however several unavoidable differences and these are clearly documented in the "Notes" section below.

<p>Most of the options in the original mast are given in ACD as "advanced" or
"additional" options. -options must be specified on the command-line in order 
to be prompted for a value for "additional" options but "advanced" options 
will never be prompted for.  

<p> Please note that one only of -stdin or -d should be specified.  If you set both, then -d will be used.   This behaviour could have been enforced at the level of the ACD file by using an ACD select: or list: type but this would have been inconsistent with the original meme, which has two separate options.
<p>

<table CELLSPACING=0 CELLPADDING=3 BGCOLOR="#f5f5ff" ><tr><td>
<pre>
   Standard (Mandatory) qualifiers:
  [-mfile]             infile     If -d <database> is not given, MAST looks
                                  for database specified inside of <mfile>.
   -ev                 float      [10] Print results for sequences with
                                  E-value (Any numeric value)
   -mt                 float      [0.0001] Show motif matches with p-value <
                                  mt (Any numeric value)
  [-outfile]           outfile    [*.emast] MAST program output file

   Additional (Optional) qualifiers:
   -d                  infile     If -d <database> is not given, MAST looks
                                  for database specified inside of <mfile>.
   -a                  infile     Input file <mfile> is assumed to contain
                                  motifs in the format output by
                                  bin/make_logodds and <a> is their alphabet;
                                  -d <database> or -stdin must be specified
                                  when this option is used.
   -bfile              infile     The random model uses the letter frequencies
                                  given in <bfile> instead of the
                                  non-redundant database frequencies. The
                                  format of <bfile> is the same as that for
                                  the MEME -bfile opton; see the MEME
                                  documentation for details. Sample files are
                                  given in directory tests: tests/nt.freq and
                                  tests/na.freq in the MEME distribution.)
   -smax               integer    [-1] Print results for no more than <smax>
                                  sequences (Any integer value)
   -stdin              boolean    [N] The default is to read the database
                                  specified inside <mfile>.
   -text               boolean    [N] Default is hypertext (HTML) format
   -dna                boolean    [N] Translate DNA sequences to protein
   -comp               boolean    [N] The random model uses the letter
                                  frequencies in the current target sequence
                                  instead of the non-redundant database
                                  frequencies. This causes p-values and
                                  E-values to be compensated individually for
                                  the actual composition of each sequence in
                                  the database. This option can increase
                                  search time substantially due to the need to
                                  compute a different score distribution for
                                  each high-scoring sequence.
   -rank               integer    [-1] Print results starting with <rank> best
                                  (Any integer value)
   -best               boolean    [N] Include only the best motif in diagrams
   -remcorr            boolean    [N] Remove highly correlated motifs from
                                  query
   -brief              boolean    [N] Brief output: do not print
                                  documentation.
   -b                  boolean    [N] Print only sections I and II
   -nostatus           boolean    [N] Do not print progress report
   -hitlist            boolean    [N] If you specify the -hitlist switch to
                                  MAST, the motif 'diagram' takes the form of
                                  a comma separated list of motif occurrences
                                  ('hits'). Each 'hit' has the format:
                                  <strand><motif> <start> <end> <p-value>
                                  where <strand> is the strand (+ or - for
                                  DNA, blank for protein), <motif> is the
                                  motif number, <start> is the starting
                                  position of the hit, <end> is the ending
                                  position of the hit, and <p-value> is the
                                  position p-value of the hit.

   Advanced (Unprompted) qualifiers:
   -c                  integer    [-1] Only use the first <c> motifs (Any
                                  integer value)
   -sep                boolean    [N] Score reverse complement DNA strand as a
                                  separate sequence
   -norc               boolean    [N] Do not score reverse complement DNA
                                  strand
   -w                  boolean    [N] Show weak matches (mt<p-value<mt*10) in
                                  angle brackets
   -seqp               boolean    [N] The default is to use POSITION p-values.
   -mf                 string     Print <mf> as motif file name. (Any string
                                  is accepted)
   -df                 string     Print <df> as database name. (Any string is
                                  accepted)
   -minseqs            integer    [-1] Lower bound on number of sequences in
                                  db (Any integer value)
   -mev                float      [-1] Use only motifs with E-values less than
                                  <mev> (Any numeric value)
   -m                  integer    [-1] Overrides value set by using -mev. (Any
                                  integer value)
   -diag               string     See on-line documentation for a valid
                                  example. (Any string is accepted)

   Associated qualifiers:

   "-outfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write standard output
   -filter             boolean    Read standard input, write standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages

</pre>
</td></tr></table>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Standard (Mandatory) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>[-mfile]<br>(Parameter 1)</td>
<td>If -d &lt;database&gt; is not given, MAST looks for database specified inside of &lt;mfile&gt;.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-ev</td>
<td>Print results for sequences with E-value</td>
<td>Any numeric value</td>
<td>10</td>
</tr>

<tr>
<td>-mt</td>
<td>Show motif matches with p-value &lt; mt</td>
<td>Any numeric value</td>
<td>0.0001</td>
</tr>

<tr>
<td>[-outfile]<br>(Parameter 2)</td>
<td>MAST program output file</td>
<td>Output file</td>
<td><i>&lt;*&gt;</i>.emast</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Additional (Optional) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-d</td>
<td>If -d &lt;database&gt; is not given, MAST looks for database specified inside of &lt;mfile&gt;.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-a</td>
<td>Input file &lt;mfile&gt; is assumed to contain motifs in the format output by bin/make_logodds and &lt;a&gt; is their alphabet; -d &lt;database&gt; or -stdin must be specified when this option is used.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-bfile</td>
<td>The random model uses the letter frequencies given in &lt;bfile&gt; instead of the non-redundant database frequencies. The format of &lt;bfile&gt; is the same as that for the MEME -bfile opton; see the MEME documentation for details. Sample files are given in directory tests: tests/nt.freq and tests/na.freq in the MEME distribution.)</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr>
<td>-smax</td>
<td>Print results for no more than &lt;smax&gt; sequences</td>
<td>Any integer value</td>
<td>-1</td>
</tr>

<tr>
<td>-stdin</td>
<td>The default is to read the database specified inside &lt;mfile&gt;.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-text</td>
<td>Default is hypertext (HTML) format</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-dna</td>
<td>Translate DNA sequences to protein</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-comp</td>
<td>The random model uses the letter frequencies in the current target sequence instead of the non-redundant database frequencies. This causes p-values and E-values to be compensated individually for the actual composition of each sequence in the database. This option can increase search time substantially due to the need to compute a different score distribution for each high-scoring sequence.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-rank</td>
<td>Print results starting with &lt;rank&gt; best</td>
<td>Any integer value</td>
<td>-1</td>
</tr>

<tr>
<td>-best</td>
<td>Include only the best motif in diagrams</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-remcorr</td>
<td>Remove highly correlated motifs from query</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-brief</td>
<td>Brief output: do not print documentation.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-b</td>
<td>Print only sections I and II</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-nostatus</td>
<td>Do not print progress report</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-hitlist</td>
<td>If you specify the -hitlist switch to MAST, the motif 'diagram' takes the form of a comma separated list of motif occurrences ('hits'). Each 'hit' has the format: &lt;strand&gt;&lt;motif&gt; &lt;start&gt; &lt;end&gt; &lt;p-value&gt; where &lt;strand&gt; is the strand (+ or - for DNA, blank for protein), &lt;motif&gt; is the motif number, &lt;start&gt; is the starting position of the hit, &lt;end&gt; is the ending position of the hit, and &lt;p-value&gt; is the position p-value of the hit.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=2>Advanced (Unprompted) qualifiers</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr>
<td>-c</td>
<td>Only use the first &lt;c&gt; motifs</td>
<td>Any integer value</td>
<td>-1</td>
</tr>

<tr>
<td>-sep</td>
<td>Score reverse complement DNA strand as a separate sequence</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-norc</td>
<td>Do not score reverse complement DNA strand</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-w</td>
<td>Show weak matches (mt&lt;p-value&lt;mt*10) in angle brackets</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-seqp</td>
<td>The default is to use POSITION p-values.</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr>
<td>-mf</td>
<td>Print &lt;mf&gt; as motif file name.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-df</td>
<td>Print &lt;df&gt; as database name.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

<tr>
<td>-minseqs</td>
<td>Lower bound on number of sequences in db</td>
<td>Any integer value</td>
<td>-1</td>
</tr>

<tr>
<td>-mev</td>
<td>Use only motifs with E-values less than &lt;mev&gt;</td>
<td>Any numeric value</td>
<td>-1</td>
</tr>

<tr>
<td>-m</td>
<td>Overrides value set by using -mev.</td>
<td>Any integer value</td>
<td>-1</td>
</tr>

<tr>
<td>-diag</td>
<td>See on-line documentation for a valid example.</td>
<td>Any string is accepted</td>
<td><i>An empty string is accepted</i></td>
</tr>

</table>

<!--
DON'T WRITE ANYTHING HERE.
IT IS DONE FOR YOU.
-->








<H2>
    Input file format
</H2>


<!-- 
        This includes example input file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null input, etc. 
   -->


<p>
<pre>



<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: ex1.html</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
&lt;!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"&gt;
&lt;HTML&gt;
&lt;HEAD&gt;
&lt;meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"&gt;
&lt;TITLE&gt;MEME&lt;/TITLE&gt;
&lt;STYLE type="text/css"&gt;
  TD.invisible { color: '#D5F0FF'; }
  TD.c0 { background: aqua; color: black; }
  TD.cw0 { background: aqua; color: black; font: 50% sans-serif; }
  TD.c1 { background: blue; color: white; }
  TD.cw1 { background: blue; color: white; font: 50% sans-serif; }
  TD.c2 { background: red; color: white; }
  TD.cw2 { background: red; color: white; font: 50% sans-serif; }
  TD.c3 { background: fuchsia; color: black; }
  TD.cw3 { background: fuchsia; color: black; font: 50% sans-serif; }
  TD.c4 { background: yellow; color: black; }
  TD.cw4 { background: yellow; color: black; font: 50% sans-serif; }
  TD.c5 { background: lime; color: black; }
  TD.cw5 { background: lime; color: black; font: 50% sans-serif; }
  TD.c6 { background: teal; color: white; }
  TD.cw6 { background: teal; color: white; font: 50% sans-serif; }
  TD.c7 { background: #444444; color: white; }
  TD.cw7 { background: #444444; color: white; font: 50% sans-serif; }
  TD.c8 { background: green; color: white; }
  TD.cw8 { background: green; color: white; font: 50% sans-serif; }
  TD.c9 { background: silver; color: black; }
  TD.cw9 { background: silver; color: black; font: 50% sans-serif; }
  TD.c10 { background: purple; color: white; }
  TD.cw10 { background: purple; color: white; font: 50% sans-serif; }
  TD.c11 { background: olive; color: black; }
  TD.cw11 { background: olive; color: black; font: 50% sans-serif; }
  TD.c12 { background: navy; color: white; }
  TD.cw12 { background: navy; color: white; font: 50% sans-serif; }
  TD.c13 { background: maroon; color: white; }
  TD.cw13 { background: maroon; color: white; font: 50% sans-serif; }
  TD.c14 { background: black; color: white; }
  TD.cw14 { background: black; color: white; font: 50% sans-serif; }
  TD.c15 { background: white; color: black; }
  TD.cw15 { background: white; color: black; font: 50% sans-serif; }
  B.red { color: red; }
  TD.red { color: red; }
  TH.red { color: red; }
  B.blue { color: blue; }
  TD.blue { color: blue; }
  TH.blue { color: blue; }
  B.orange { color: orange; }
  TD.orange { color: orange; }
  TH.orange { color: orange; }
  B.green { color: green; }
  TD.green { color: green; }


<font color=red>  [Part of this file has been deleted for brevity]</font>

for use by database search programs such as MAST.  This matrix is a
log-odds matrix calculated
by taking the log (base 2) of the ratio &lt;TT&gt;p/f&lt;/TT&gt; at each position in
the motif where &lt;TT&gt;p&lt;/TT&gt; is the probability of a particular letter at that
position in the motif, and &lt;TT&gt;f&lt;/TT&gt; is the background frequency of the
letter (given in the &lt;A HREF=#command_doc&gt;command line summary&lt;/A&gt; section.)
Each entry in the matrix is multiplied by 100 and rounded to the nearest
integer before printing.
This is the same matrix that is used above in computing the &lt;I&gt;p&lt;/I&gt;-values
of the occurrences of the motif in the &lt;A HREF=#sites_doc2&gt;Occurrences of the
Motif&lt;/A&gt; and &lt;A HREF=#diagrams_doc2&gt;Block Diagrams of Motif Occurrences&lt;/A&gt;
sections.

The scoring matrix is printed "sideways"--columns
correspond to the letters in the alphabet (in the same order as shown in
the simplified motif) and rows corresponding to the positions of the motif,
position one first.  The scoring matrix is preceded by a line starting with
"log-odds matrix:" and containing the length of the alphabet, width
of the motif, number of characters in the training set and a scoring 
threshold.

&lt;P&gt;
&lt;LI&gt; &lt;A NAME=pspm_doc2 HREF=#pspm1&gt;&lt;H4&gt;
Position-Specific Probability Matrix&lt;/H4&gt;&lt;/A&gt;

The motif itself is a position-specific probability matrix giving,
for each position in the pattern, the observed frequency 
("probability") of each possible letter.  
The probability matrix is printed "sideways"--columns
correspond to the letters in the alphabet (in the same order as shown in
the simplified motif) and rows corresponding to the positions of the motif,
position one first.
The motif is preceded by a line starting with
"letter-probability matrix:" and containing the length of the alphabet, width
of the motif, number of occurrences of the motif, and the E-value of the
motif.
&lt;p&gt;
&lt;b&gt;Note:&lt;/b&gt; Earlier versions
of MEME gave the posterior probabilities--the probability after applying
a prior on letter frequencies--rather than the observed frequencies.
These versions of MEME also gave the number of &lt;I&gt;possible&lt;/I&gt;
positions for the motif rather than the actual number of occurrences.
The output from these earlier versions of MEME can be distinguished
by "n=" rather than "nsites=" in the line preceding the matrix.

&lt;/UL&gt;
&lt;HR&gt;&lt;TABLE SUMMARY='buttons' ALIGN=LEFT CELLSPACING=0&gt;&lt;TR&gt;
  &lt;TD BGCOLOR='#DDDDFF'&gt;&lt;A HREF="#top_buttons"&gt;&lt;B&gt;Go to top&lt;/B&gt;&lt;/A&gt;&lt;/TABLE&gt;&lt;BR&gt;
&lt;/FORM&gt;
&lt;/BODY&gt;
&lt;/HTML&gt;
</td></tr></table><p>

</pre>

<h3>  MOTIF FORMAT  </h3>
  
  MAST can search using (multiple) motifs contained in 
<ul>  
<li>      a MEME output file, 
<li>      a GCG profile file, 
<li>      two or more GCG profile filess concatenated together, or 
<li>      a file with the following format. 
</ul>  

<h3>                    Motif file format </h3>
<pre>  
       ALPHABET= alphabet
       log-odds matrix: alength= alength w= w
       row_1
       row_2
       ...
       row_w  
</pre>  
  
  
<p>      A motif is represented by a position-dependent scoring matrix. 
<p>      A scoring matrix is preceded by a line starting with the words
      log-odds matrix: and specifying alength, the length of
      the alphabet (number of columns in the scoring matrix), and the w, the
      width of the motif (number of rows in the scoring matrix). 
<p>      The following w lines (no blank lines allowed) contain the rows of the
      scoring matrix. Row i, column j of the matrix gives the score for the j-th
      letter in alphabet appearing at position i in an occurrence of the
      motif. 

<p>      The spaces after the equals signs and the colon are required. 
<p>      The number of letters in alphabet must equal alength. 
<p>      Any number of additional motifs may follow the first one. 
<p>      The motif file must contain a line starting with 
  <pre>
              ALPHABET= 
</pre>  
      followed by alphabet, a list containing the letters used in the motifs. 

<p>      The order of the letters in alphabet must be the same as the order of the
      columns of scores in the motifs. The order need not be alphabetical
      and case does not matter, but there should be no spaces in alphabet.
<p>      The letters in alphabet must be a subset of either the IUB/IUPAC DNA
      (ABCDGHKMNRSTUVWY) or protein
      (ABCDEFGHIKLMNPQRSTUVWXYZ) alphabets. DNA alphabets
      must contain at least the letters ACGT. Protein alphabets must contain
      at least the letters ACDEFGHIKLMNPQRSTVWY. All other letters in
      the alphabets are optional. If any of the optional letters are missing 
      from alphabet, MAST automatically generates scores for them by taking the
      weighted average of the scores for the letters which the missing letter
      could match. (The weights are the frequencies of the replaced letters in
      the appropriate non-redundant database.) Replacements for the
      optional letters are given in the following table. 
  
<h3>             LETTERS MATCHED BY OPTIONAL LETTERS </h3>
<pre>
      =================================================
      optional          matches 
      letter      DNA             protein 
      =================================================
       B          CGT             DN 
       D          AGT
       H          ACT
       K          GT
       M          AC
       N          ACGT
       R          AG
       S          CG
       U          T               ACDEFGHIKLMNPQRSTVWY 
       V          CAG
       W          AT
       X                          ACDEFGHIKLMNPQRSTVWY 
       Y          CT
       Z                          EQ 
       *          ACGT            ACDEFGHIKLMNPQRSTVWY
       -          ACGT            ACDEFGHIKLMNPQRSTVWY
      =================================================
</pre>

<h3>  EXAMPLE <h3>
  
  Here is an example of a DNA motif file that contains two motifs. 
  
<h3>                    Sample motif file <h3>
<pre>  
          ALPHABET= ACGT
          log-odds matrix: alength= 4 w= 9
           -4.275  -0.182  -4.195   1.408
           -4.296  -1.487   1.880  -0.816
           -2.160  -1.492  -4.171   1.474
           -0.810  -4.076   1.872  -2.164
            1.537  -1.487  -4.195  -4.205
            0.113   0.340  -0.237  -0.209
           -0.454   0.923   0.390  -0.834
           -1.336  -0.082   0.905   0.100
            0.674  -4.183   0.130  -0.201
          log-odds matrix: alength= 4 w= 6
           -2.032   0.324   1.371  -0.781
           -0.409   0.560  -0.250   0.119
           -4.274  -0.519  -0.260   1.167
           -2.188   2.300  -4.191  -2.465
            1.265  -4.111  -0.267  -2.180
           -1.977   2.158  -1.661  -2.071 
</pre>
  
  
  In the example above, because the order of the letters in alphabet is
  ACGT, the first column of each motif gives the scores for the letter A at each
  position in the motif, the second column gives the scores for C and so forth.
  
<p>  Note: If -d <database> is not given, MAST looks for database
  	specified inside of < mfile >
  
<p>  Creates file (unless [-stdout] given) after stripping ".html" from the end of
  < mfile >:
<p>  	mast.< mfile >[.< database >][.c< count >][.m< motif >]+[.rank< rank >][.ev< ev >][.mt< mt >][.b]






<H2>
    Output file format
</H2>

<!-- 
        This includes example output file formats.
        This should be a detailed description and example - assume
        someone will want to parse this file and will want to know what
        happens in unusual cases - null output, errors etc. 

        If you wish to include the standard description of the avalable
	report formats, use:
include file="inc/reportformats.ihtml"
   -->

<p>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: ex1.out</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&lt;!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"&gt;
&lt;HTML&gt;
&lt;HEAD&gt;
&lt;meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"&gt;
&lt;TITLE&gt;MAST&lt;/TITLE&gt;
&lt;STYLE type="text/css"&gt;
  TD.invisible { color: '#D5F0FF'; }
  TD.c0 { background: aqua; color: black; }
  TD.cw0 { background: aqua; color: black; font: 50% sans-serif; }
  TD.c1 { background: blue; color: white; }
  TD.cw1 { background: blue; color: white; font: 50% sans-serif; }
  TD.c2 { background: red; color: white; }
  TD.cw2 { background: red; color: white; font: 50% sans-serif; }
  TD.c3 { background: fuchsia; color: black; }
  TD.cw3 { background: fuchsia; color: black; font: 50% sans-serif; }
  TD.c4 { background: yellow; color: black; }
  TD.cw4 { background: yellow; color: black; font: 50% sans-serif; }
  TD.c5 { background: lime; color: black; }
  TD.cw5 { background: lime; color: black; font: 50% sans-serif; }
  TD.c6 { background: teal; color: white; }
  TD.cw6 { background: teal; color: white; font: 50% sans-serif; }
  TD.c7 { background: #444444; color: white; }
  TD.cw7 { background: #444444; color: white; font: 50% sans-serif; }
  TD.c8 { background: green; color: white; }
  TD.cw8 { background: green; color: white; font: 50% sans-serif; }
  TD.c9 { background: silver; color: black; }
  TD.cw9 { background: silver; color: black; font: 50% sans-serif; }
  TD.c10 { background: purple; color: white; }
  TD.cw10 { background: purple; color: white; font: 50% sans-serif; }
  TD.c11 { background: olive; color: black; }
  TD.cw11 { background: olive; color: black; font: 50% sans-serif; }
  TD.c12 { background: navy; color: white; }
  TD.cw12 { background: navy; color: white; font: 50% sans-serif; }
  TD.c13 { background: maroon; color: white; }
  TD.cw13 { background: maroon; color: white; font: 50% sans-serif; }
  TD.c14 { background: black; color: white; }
  TD.cw14 { background: black; color: white; font: 50% sans-serif; }
  TD.c15 { background: white; color: black; }
  TD.cw15 { background: white; color: black; font: 50% sans-serif; }
  B.red { color: red; }
  TD.red { color: red; }
  TH.red { color: red; }
  B.blue { color: blue; }
  TD.blue { color: blue; }
  TH.blue { color: blue; }
  B.orange { color: orange; }
  TD.orange { color: orange; }
  TH.orange { color: orange; }
  B.green { color: green; }
  TD.green { color: green; }


<font color=red>  [Part of this file has been deleted for brevity]</font>


&lt;HR&gt;
&lt;A NAME=a17&gt;&lt;/A&gt;ilv &lt;TABLE SUMMARY='buttons' ALIGN=LEFT CELLSPACING=0&gt;&lt;TR&gt;
  &lt;TD BGCOLOR='#DDDD88'&gt;&lt;A HREF='#s17'&gt;S&lt;/A&gt;
  &lt;TD BGCOLOR='#DDFFDD'&gt;&lt;A HREF='#d17'&gt;D&lt;/A&gt;
  &lt;TD BGCOLOR='#FFFFFF'&gt;&lt;A HREF='#bh'&gt;?&lt;/A&gt;&lt;/TABLE&gt;&lt;BR CLEAR=LEFT&gt;
  &lt;BR&gt;
  LENGTH = 105  COMBINED P-VALUE = 6.96e-02  E-VALUE =      1.3&lt;BR&gt;
  DIAGRAM: 105&lt;PRE&gt;


&lt;/PRE&gt;

&lt;HR&gt;
&lt;A NAME=a18&gt;&lt;/A&gt;trn9cat &lt;TABLE SUMMARY='buttons' ALIGN=LEFT CELLSPACING=0&gt;&lt;TR&gt;
  &lt;TD BGCOLOR='#DDDD88'&gt;&lt;A HREF='#s18'&gt;S&lt;/A&gt;
  &lt;TD BGCOLOR='#DDFFDD'&gt;&lt;A HREF='#d18'&gt;D&lt;/A&gt;
  &lt;TD BGCOLOR='#FFFFFF'&gt;&lt;A HREF='#bh'&gt;?&lt;/A&gt;&lt;/TABLE&gt;&lt;BR CLEAR=LEFT&gt;
  &lt;BR&gt;
  LENGTH = 105  COMBINED P-VALUE = 2.43e-01  E-VALUE =      4.4&lt;BR&gt;
  DIAGRAM: 105&lt;PRE&gt;&lt;/PRE&gt;

&lt;A NAME=debug&gt;&lt;/A&gt;&lt;HR&gt;&lt;CENTER&gt;&lt;H3&gt;Debugging Information&lt;/H3&gt;&lt;/CENTER&gt;&lt;HR&gt;
&lt;PRE&gt;


CPU: emboss6.ebi.ac.uk
Time 0.001999 secs.

mast ../../data/memenew/ex1.html -ev 10.000000 -mt 0.000100
&lt;/PRE&gt;
&lt;A NAME=bh&gt;&lt;/A&gt;
&lt;A NAME=sbh&gt;&lt;/A&gt;
&lt;A NAME=dbh&gt;&lt;/A&gt;
&lt;A NAME=abh&gt;&lt;/A&gt;
&lt;HR&gt;&lt;CENTER&gt;&lt;H3&gt;Button Help&lt;/H3&gt;&lt;/CENTER&gt;&lt;HR&gt;
&lt;TABLE SUMMARY='buttons' ALIGN=LEFT CELLSPACING=0&gt;&lt;TR&gt;
  &lt;TD BGCOLOR='#DDDDFF'&gt;&lt;A HREF='#bh'&gt;E&lt;/A&gt;&lt;/TABLE&gt;Links to Entrez database at &lt;A HREF='http://www.ncbi.nlm.nih.gov'&gt;NCBI&lt;/A&gt; &lt;BR CLEAR=LEFT&gt;
&lt;TABLE SUMMARY='buttons' ALIGN=LEFT CELLSPACING=0&gt;&lt;TR&gt;
  &lt;TD BGCOLOR='#DDDD88'&gt;&lt;A HREF='#sbh'&gt;S&lt;/A&gt;&lt;/TABLE&gt;Links to sequence scores (&lt;A HREF='#sec_i'&gt;section I&lt;/A&gt;) &lt;BR CLEAR=LEFT&gt;
&lt;TABLE SUMMARY='buttons' ALIGN=LEFT CELLSPACING=0&gt;&lt;TR&gt;
  &lt;TD BGCOLOR='#DDFFDD'&gt;&lt;A HREF='#dbh'&gt;D&lt;/A&gt;&lt;/TABLE&gt;Links to motif diagrams (&lt;A HREF='#sec_ii'&gt;section II&lt;/A&gt;) &lt;BR CLEAR=LEFT&gt;
&lt;TABLE SUMMARY='buttons' ALIGN=LEFT CELLSPACING=0&gt;&lt;TR&gt;
  &lt;TD BGCOLOR='#FFDDDD'&gt;&lt;A HREF='#abh'&gt;A&lt;/A&gt;&lt;/TABLE&gt;Links to sequence/motif annotated alignments (&lt;A HREF='#sec_iii'&gt;section III&lt;/A&gt;) &lt;BR CLEAR=LEFT&gt;
&lt;TABLE SUMMARY='buttons' ALIGN=LEFT CELLSPACING=0&gt;&lt;TR&gt;
  &lt;TD BGCOLOR='#FFFFFF'&gt;&lt;A HREF='#bh'&gt;?&lt;/A&gt;&lt;/TABLE&gt;This information &lt;BR CLEAR=LEFT&gt;

&lt;HR&gt;&lt;TABLE SUMMARY='buttons' ALIGN=LEFT CELLSPACING=0&gt;&lt;TR&gt;
  &lt;TD BGCOLOR='#DDDDFF'&gt;&lt;A HREF='#top_buttons'&gt;&lt;B&gt;Go to top&lt;/B&gt;&lt;/A&gt;&lt;/TABLE&gt;&lt;BR&gt;
&lt;/BODY&gt;
&lt;/HTML&gt;
</pre>
</td></tr></table><p>
<p><h3>File: crp0.s</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
&gt;ce1cg
TAATGTTTGTGCTGGTTTTTGTGGCATCGGGCGAGAATAGCGCGTGGTGTGAAAGACTGTTTTTTTGATCGTTTTCACAA
AAATGGAAGTCCACAGTCTTGACAG
&gt;ara
GACAAAAACGCGTAACAAAAGTGTCTATAATCACGGCAGAAAAGTCCACATTGATTATTTGCACGGCGTCACACTTTGCT
ATGCCATAGCATTTTTATCCATAAG
&gt;bglr1
ACAAATCCCAATAACTTAATTATTGGGATTTGTTATATATAACTTTATAAATTCCTAAAATTACACAAAGTTAATAACTG
TGAGCATGGTCATATTTTTATCAAT
&gt;crp
CACAAAGCGAAAGCTATGCTAAAACAGTCAGGATGCTACAGTAATACATTGATGTACTGCATGTATGCAAAGGACGTCAC
ATTACCGTGCAGTACAGTTGATAGC
&gt;cya
ACGGTGCTACACTTGTATGTAGCGCATCTTTCTTTACGGTCAATCAGCAAGGTGTTAAATTGATCACGTTTTAGACCATT
TTTTCGTCGTGAAACTAAAAAAACC
&gt;deop2
AGTGAATTATTTGAACCAGATCGCATTACAGTGATGCAAACTTGTAAGTAGATTTCCTTAATTGTGATGTGTATCGAAGT
GTGTTGCGGAGTAGATGTTAGAATA
&gt;gale
GCGCATAAAAAACGGCTAAATTCTTGTGTAAACGATTCCACTAATTTATTCCATGTCACACTTTTCGCATCTTTGTTATG
CTATGGTTATTTCATACCATAAGCC
&gt;ilv
GCTCCGGCGGGGTTTTTTGTTATCTGCAATTCAGTACAAAACGTGATCAACCCCTCAATTTTCCCTTTGCTGAAAAATTT
TCCATTGTCTCCCCTGTAAAGCTGT
&gt;lac
AACGCAATTAATGTGAGTTAGCTCACTCATTAGGCACCCCAGGCTTTACACTTTATGCTTCCGGCTCGTATGTTGTGTGG
AATTGTGAGCGGATAACAATTTCAC
&gt;male
ACATTACCGCCAATTCTGTAACAGAGATCACACAAAGCGACGGTGGGGCGTAGGGGCAAGGAGGATGGAAAGAGGTTGCC
GTATAAAGAAACTAGAGTCCGTTTA
&gt;malk
GGAGGAGGCGGGAGGATGAGAACACGGCTTCTGTGAACTAAACCGAGGTCATGTAAGGAATTTCGTGATGTTGCTTGCAA
AAATCGTGGCGATTTTATGTGCGCA
&gt;malt
GATCAGCGTCGTTTTAGGTGAGTTGTTAATAAAGATTTGGAATTGTGACACAGTGCAAATTCAGACACATAAAAAAACGT
CATCGCTTGCATTAGAAAGGTTTCT
&gt;ompa
GCTGACAAAAAAGATTAAACATACCTTATACAAGACTTTTTTTTCATATGCCTGACGGAGTTCACACTTGTAAGTTTTCA
ACTACGTTGTAGACTTTACATCGCC
&gt;tnaa
TTTTTTAAACATTAAAATTCTTACGTAATTTATAATCTTTAAAAAAAGCATTTAATATTGCTCCCCGAACGATTGTGATT
CGATTCACATTTAAACAATTTCAGA
&gt;uxu1
CCCATGAGAGTGAAATTGTTGTGATGTGGTTAACCCAATTAGAATTCGGGATTGACATGTCTTACCAAAAGGTAGAACTT
ATACGCCATCTCATCCGATGCAAGC
&gt;pbr322
CTGGCTTAACTATGCGGCATCAGAGCAGATTGTACTGAGAGTGCACCATATGCGGTGTGAAATACCGCACAGATGCGTAA
GGAGAAAATACCGCATCAGGCGCTC
&gt;trn9cat
CTGTGACGGAAGATCACTTCGCAGAATAAATAAATCCTGGTGTCCCTGTTGATACCGGGAAGCCCTGGGCCAACTTTTGG
CGAAAATGAGACGTTGATCGGCACG
&gt;tdc
GATTTTTATACTTTAACTTGTTGATATTTAAAGGTATTTAATTGTAATAACGATACTCTGGAAAGTATTGAAAGTTAATT
TGTGAGTGGTCGCACATATCCTGTT
</pre>
</td></tr></table><p>


<p>  MAST outputs a file containing:
<ul>  
<li>      * the version of MAST and the date it was built, 
<li>      * the reference to cite if you use MAST in your research, 
<li>       * a description of the database and motifs used in the search, 
<li>       * an explanation of the results,
<li>       * high-scoring sequences--sequences matching the group of motifs above a stated level of statistical significance, 
<li>       * motif diagrams showing the order and spacing of occurrences of the motifs in the high-scoring sequences and 
<li>       * annotated sequences showing the positions and p-values of all motif occurrences in each of the high-scoring sequences. 
</ul>  

<p>  Each section of the results file contains an explanation of how to interpret
  them. 
  
<h3>    Match Scores <h3>
  
  The match score of a motif to a position in a sequence is the sum of the
  score from each column of the position-dependent scoring matrix
  corresponding to the letter at that position in the sequence. For example, if
  the sequence is 

<pre>  
  TAATGTTGGTGCTGGTTTTTGTGGCATCGGGCGAGAATAGCGC
     ========
 </pre>

  and the motif is represented by the position-dependent scoring matrix (where
  each row of the matrix corresponds to a position in the motif) 

<pre>  
  =========|=================================
  POSITION |   A        C        G        T
  =========|=================================
    1      | 1.447    0.188   -4.025   -4.095 
    2      | 0.739    1.339   -3.945   -2.325 
    3      | 1.764   -3.562   -4.197   -3.895 
    4      | 1.574   -3.784   -1.594   -1.994 
    5      | 1.602   -3.935   -4.054   -1.370 
    6      | 0.797   -3.647   -0.814    0.215 
    7      |-1.280    1.873   -0.607   -1.933 
    8      |-3.076    1.035    1.414   -3.913 
  =========|=================================
</pre>  

  then the match score of the fourth position in the sequence (underlined)
  would be found by summing the score for T in position 1, G in position 2 and
  so on until G in position 8. So the match score would be 
  
<pre>
    score = -4.095 + -3.945 + -3.895 + -1.994
            + -4.054 + -0.814 + -1.933 + 1.414 
          = -19.316
  </pre>
  The match scores for other positions in the sequence are calculated in the
  same way. Match scores are only calculated if the match completely fits within
  the sequence. Match scores are not calculated if the motif would overhang
  either end of the sequence. 
  
<h3>    P-values </h3>
  
  MAST reports all matches of a sequence to a motif or group of motifs in terms
  of the p-value of the match. MAST considers the p-values of four types of
  events: 
<ul>
<li>  position p-value: the match of a single position within a sequence to
      	a given motif, 
<li>      sequence p-value: the best match of any position within a sequence
      	to a given motif, 
<li>      combined p-value: the combined best matches of a sequence to a
      	group of motifs, and 
<li>      E-value: observing a combined p-value at least as small in a random
      	database of the same size. 
</ul>
  
  All p-values are based on a random sequence model that assumes each
  position in a random sequence is generated according to the average letter
  frequencies of all sequences in the the appropriate (peptide or nucleotide)
  non-redundant database (ftp://ncbi.nlm.nih.gov/blast/db/) on September 22,
  1996.  This can be overridden in two ways:
  
<h3>  	1) -bfile < bfile > </h3>
  	The random model uses the letter frequencies given in < bfile > 
  	instead of the non-redundant database frequencies.
  	The format of < bfile > is the same as that for the MEME -bfile opton; 
  	see the MEME documentation for details.  Sample files are given in 
  	directory tests: tests/nt.freq and tests/na.freq.) 
  	
<h3>  	2) -comp </h3>
  	The random model uses the letter frequencies in the current target
  	sequence instead of the non-redundant database frequencies.  This
  	causes p-values and E-values to be compensated individually for the 
  	actual composition of each sequence in the database.  This option
  	can increase search time substantially due to the need to compute
  	a different score distribution for each high-scoring sequence.
  
  
<h3>      Position p-value </h3>
  
      The p-value of a match of a given position within a sequence to a
      motif is defined as the probability of a randomly selected position in a
      randomly generated sequence having a match score at least as large
      as that of the given position. 
  
<h3>      Sequence p-value </h3>
  
      The p-value of a match of a sequence to a motif is defined as the
      probability of a randomly generated sequence of the same length
      having a match score at least as large as the largest match score of
      any position in the sequence. 
  
<h3>      Combined p-value </h3>
  
      The p-value of a match of a sequence to a group of motifs is defined
      as the probability of a randomly generated sequence of the same
      length having sequence p-values whose product is at least as small
      as the product of the sequence p-values of the matches of the motifs
      to the given sequence. 
  
<h3>      E-value </h3>
  
      The E-value of the match of a sequence in a database to a a group
      of motifs is defined as the expected number of sequences in a random
      database of the same size that would match the motifs as well as the
      sequence does and is equal to the combined p-value of the sequence
      times the number of sequences in the database. 
  
<h3>    High-scoring Sequences </h3>
  
  MAST lists the names and part of the descriptive text of all sequences
  whose E-value is less than E. Sequences shorter than one or more of the
  motifs are skipped. The sequences are sorted by increasing E-value. The
  value of E is set to 10 for the WEB server but is user-selectable in the
  down-loadable version of MAST. 
  
<h3>    Motif Diagrams </h3>
  
  Motif diagrams show the order and spacing of non-overlapping matches to
  the motifs in each high-scoring sequence. Motif occurrences are determined
  based on the position p-value of matches to the motif. Strong matches
  (p-value < M) are shown in square brackets (`[ ]'), weak matches (M <
  p-value < M × 10) are shown in angle brackets (`< >') and the length of
  non-motif sequence ("spacer") is shown between dashes (`-'). For example, 
<pre>  
          27-[3]-44-< 4 >-99-[1]-7
</pre>  
  shows an initial spacer of length 27, followed by a strong match to motif 3, a
  spacer of length 44, a weak match to motif 4, a spacer of length 99, a strong
  match to motif 1 and a final non-motif sequence of length 7. The value of M is
  0.0001 for the WEB server but is user-selectable in the down-loadable
  version of MAST. 
  
<p>
  Note: If you specify the -hit_list switch to MAST, the motif "diagram" takes the form
  of a comma separated list of motif occurrences ("hits").  Each "hit" has the format:
  	< strand >< motif > < start > < end > < p-value >
  where 
<ul>
<li>          < strand >        is the strand (+ or - for DNA, blank for protein),
<li>          < motif >         is the motif number,
<li>          < start >         is the starting position of the hit,
<li>          < end >           is the ending position of the hit, and
<li>          < p-value >       is the position p-value of the hit.
</ul>  


<h3>    Annotated Sequences </h3>
  
  MAST annotates each high-scoring sequence by printing the sequence
  along with the position and strength of all the non-overlapping motif
  occurrences. The four lines above each motif occurrence contain,
  respectively, 
<ul>  
<li>      the motif number of the occurrence, 
<li>      the position p-value of the occurence, 
<li>      the best possible match to the motif, and 
<li>      a plus sign (`+') above each letter in the occurrence that has a positive
<li>      match score to the motif. 
</ul>  
  The best possible match to a motif is the sequence of letters which would
  acheive the highest match score. 





<H2>
    Data files
</H2>

<!-- 
        Any data files used (e.g. translation table file)
        This includes example data file formats if they are
        not obvious.

	If you wish to include the standard description of what data
	files are and how to use embossdata to inspect and retrieve
	them, use:
include file="inc/localfiles.ihtml"
   -->

None.













<H2>
    Notes
</H2>

<!-- 
        Restrictions.
        Interesting behaviour.
        Useful things you can do with this program.
   -->
<H3>1. Command-line arguments</H3>
The following original MEME options are not supported:
<br><pre>
-stdout       : The output is always written to file.
-hit_list     : Use -hitlist instead. 
</pre>

<p>
The following additional options are provided:
<br><pre>
outfile       : Application output that was normally written to stdout.
</pre>



<H3>2. Installing EMBASSY MEME</H3>
The EMBASSY MEME package contains "wrapper" applications providing an EMBOSS-style interface to the applications in the original MEME package version 3.0.14 developed by Timothy L. Bailey.  Please read the file README in the EMBASSY MEME package distribution for installation instructions.



<H3>3. Installing original MEME</H3>
To use EMBASSY MEME, you will first need to download and install the original MEME package:
<pre>
WWW home:       http://meme.sdsc.edu/meme/
Distribution:   http://meme.nbcr.net/downloads/old_versions/  
</pre>
Please read the file README in the the original MEME package distribution for installation instructions.


<H3>4. Setting up MEME</H3>
For the EMBASSY MEME package to work, the directory containing the original MEME executables *must* be in your path. For example if you executables were installed to "/usr/local/meme/bin", then type:
<pre>
set path=(/usr/local/meme/bin/ $path)
rehash
</pre>

<H3> 5. Getting help </H3>
Once you have installed the original MEME, type
<pre>
meme > meme.txt 
mast > mast.txt 
</pre>
to retrieve the meme and mast documentation into text files. The same documentation is given here and in the ememe documentation.
<p>Please read the 'Notes' section below for a description of the differences between the original and EMBASSY MEME, particularly which application command line options are supported. 






<H2>
    References
</H2>

<!-- 
        Bibliography for methods used.
   -->

<p>(MEME) Timothy L. Bailey and Charles Elkan, "Fitting a mixture model by expectation maximization to discover motifs in biopolymers", Proceedings of the Second International Conference on Intelligent Systems for Molecular Biology, pp. 28-36, AAAI Press, Menlo Park, California, 1994.
<p>(MAST) Timothy L. Bailey and Michael Gribskov, "Combining evidence using p-values: application to sequence homology searches", Bioinformatics, Vol. 14, pp. 48-54, 1998. 








<H2>
    Warnings
</H2>

<!-- 
        Potentially stupid things the program will let you do.
   -->



<H2>
    Diagnostic Error Messages
</H2>

<!-- 
        Error messages specific to this program, eg:
        "FATAL xxx" - means you have not set up the xxx data using program <b>prog</b>.<p>
   -->

None.







<H2>
    Exit status
</H2>

<!-- 
        Description of the exit status for various error conditions
   -->

It always exits with status 0.








<H2>
    Known bugs
</H2>


<!-- 
        Bugs noted but not yet fixed.
   -->

None.








<!--
<H2>
    See also
</H2>
-->
<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th><th>Description</th></tr>
<tr>
<td><a href="/emboss/apps/antigenic.html">antigenic</a></td>
<td>Finds antigenic sites in proteins</td>
</tr>

<tr>
<td><a href="/emboss/apps/digest.html">digest</a></td>
<td>Protein proteolytic enzyme or reagent cleavage digest</td>
</tr>

<tr>
<td><a href="ememe.html">ememe</a></td>
<td>Motif detection</td>
</tr>

<tr>
<td><a href="/emboss/apps/epestfind.html">epestfind</a></td>
<td>Finds PEST motifs as potential proteolytic cleavage sites</td>
</tr>

<tr>
<td><a href="/emboss/apps/fuzzpro.html">fuzzpro</a></td>
<td>Protein pattern search</td>
</tr>

<tr>
<td><a href="/emboss/apps/fuzztran.html">fuzztran</a></td>
<td>Protein pattern search after translation</td>
</tr>

<tr>
<td><a href="/emboss/apps/helixturnhelix.html">helixturnhelix</a></td>
<td>Report nucleic acid binding motifs</td>
</tr>

<tr>
<td><a href="/emboss/apps/oddcomp.html">oddcomp</a></td>
<td>Find protein sequence regions with a biased composition</td>
</tr>

<tr>
<td><a href="/embassy/meme/omeme.html">omeme</a></td>
<td>Motif detection</td>
</tr>

<tr>
<td><a href="/emboss/apps/patmatdb.html">patmatdb</a></td>
<td>Search a protein sequence with a motif</td>
</tr>

<tr>
<td><a href="/emboss/apps/patmatmotifs.html">patmatmotifs</a></td>
<td>Search a PROSITE motif database with a protein sequence</td>
</tr>

<tr>
<td><a href="/emboss/apps/pepcoil.html">pepcoil</a></td>
<td>Predicts coiled coil regions</td>
</tr>

<tr>
<td><a href="/emboss/apps/preg.html">preg</a></td>
<td>Regular expression search of a protein sequence</td>
</tr>

<tr>
<td><a href="/emboss/apps/pscan.html">pscan</a></td>
<td>Scans proteins using PRINTS</td>
</tr>

<tr>
<td><a href="/emboss/apps/sigcleave.html">sigcleave</a></td>
<td>Reports protein signal cleavage sites</td>
</tr>

</table>
<!-- 
        Add any comments about other associated programs (to prepare
        data files?) that seealso doesn't find. 
   -->










<H2>
    Author(s)
</H2>

This program is an EMBOSS conversion of a program written by Sean Eddy
as part of his HMMER package.

<p>
Although we take every care to ensure that the results of the EMBOSS
version are identical to those from the original package, we recommend
that you check your inputs give the same results in both versions
before publication.

<p>
Please report all bugs in the EMBOSS version to the EMBOSS bug team,
not to the original author.
Jon Ison (jison&nbsp;&copy;&nbsp;rfcgr.mrc.ac.uk)
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK


<p>This program is an EMBASSY wrapper to a program written by Timothy L. Bailey as part of his meme package.
<p>Please report any bugs to the EMBOSS bug team in the first instance, not to Timothy L. Bailey.

<H2>
    History
</H2>

None.



<H2>
    Target users
</H2>

<!--
        For general users, requested by one user, for EMBOSS site
        maintainers, for EMBOSS developers etc.
        eg:
        "This program is intended to be used by everyone and everything,
        from naive users to embedded scripts." 
	Which is easy to include using:
   -->

This program is intended to be used by everyone and everything, from naive users to embedded scripts.











</BODY>
</HTML>

