<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: DOMAINREP documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<table align=center border=0 cellspacing=0 cellpadding=0>
<tr><td valign=top>
<A HREF="/" ONMOUSEOVER="self.status='Go to the EMBOSS home page';return true"><img border=0 src="emboss_icon.jpg" alt="" width=150 height=48></a>
</td>
<td align=left valign=middle>
<b><font size="+6">
<H2> DOMAINREP documentation
</font></b>
</td></tr>
</table>
<br>&nbsp;
<p>

<!-- END OF HEADER -->






<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<br><H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>

<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>

Reorder DCF file to identify representative structures

<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>

DOMAINREP reads a DCF file (domain classification file) and writes a DCF file containing the same data except that the order of domains within each user-specified node (e.g. family) in the hierarchy may be changed so that the representative structure of each node is given first.  The representative structure is found by using stamp and this application must be installed on the system that is running DOMAINREP and correctly configured (see 'Notes' below).  A 'domain PDB file' for each domain in the DCF file must exist in a directory specified in the stamp "pdb.directories" file (see 'Notes' below).  The domain PDB file contains coordinate data for a single domain from SCOP or CATH, in PDB format.  The input and output files are specified by the user.






<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.

     Cannot use the test data files because they might be empty or need 
     hand-editing 

     Use "<b>DOMAINREP</b> reads any normal sequence USAs." if
     appropriate.
-->
<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>
The format of the DCF file is described in the <a href="scopparse.html">SCOPPARSE documentation.</a>



<a name="input.1"></a>
<h3>Input files for usage example </h3>
<p><h3>File: all.scop2</h3>
<table width="90%"><tr><td bgcolor="#FFCCFF">
<pre>
ID   D1CS4A_
XX
EN   1CS4
XX
TY   SCOP
XX
SI   53931 CL; 54861 FO; 55073 SF; 55074 FA; 55077 DO; 55078 SO; 39418 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Ferredoxin-like
XX
SF   Adenylyl and guanylyl cyclase catalytic domain
XX
FA   Adenylyl and guanylyl cyclase catalytic domain
XX
DO   Adenylyl cyclase VC1, domain C1a
XX
OS   Dog (Canis familiaris)
XX
NC   1
XX
CN   [1]
XX
CH   A CHAIN; . START; . END;
//
ID   D1FX2A_
XX
EN   1FX2
XX
TY   SCOP
XX
SI   53931 CL; 54861 FO; 55073 SF; 55074 FA; 55081 DO; 55082 SO; 39430 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Ferredoxin-like
XX
SF   Adenylyl and guanylyl cyclase catalytic domain
XX
FA   Adenylyl and guanylyl cyclase catalytic domain
XX
DO   Receptor-type monomeric adenylyl cyclase
XX
OS   Trypanosome (Trypanosoma brucei), different isoform
XX
NC   1
XX
CN   [1]
XX


<font color=red>  [Part of this file has been deleted for brevity]</font>

XX
EN   4AT1
XX
TY   SCOP
XX
SI   53931 CL; 54861 FO; 54893 SF; 54894 FA; 54895 DO; 54896 SO; 39019 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Ferredoxin-like
XX
SF   Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain
XX
FA   Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain
XX
DO   Aspartate carbamoyltransferase
XX
OS   Escherichia coli
XX
NC   1
XX
CN   [1]
XX
CH   B CHAIN; 8 START; 100 END;
//
ID   D4AT1D1
XX
EN   4AT1
XX
TY   SCOP
XX
SI   53931 CL; 54861 FO; 54893 SF; 54894 FA; 54895 DO; 54896 SO; 39020 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Ferredoxin-like
XX
SF   Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain
XX
FA   Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain
XX
DO   Aspartate carbamoyltransferase
XX
OS   Escherichia coli
XX
NC   1
XX
CN   [1]
XX
CH   D CHAIN; 8 START; 100 END;
//
</pre>
</td></tr></table><p>


<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>
The format of the DCF file is described in the <a href="scopparse.html">SCOPPARSE documentation.</a>


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p><h3>File: all_rep1st.scop</h3>
<table width="90%"><tr><td bgcolor="#CCFFCC">
<pre>
ID   D4AT1B1
XX
EN   4AT1
XX
TY   SCOP
XX
SI   53931 CL; 54861 FO; 54893 SF; 54894 FA; 54895 DO; 54896 SO; 39019 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Ferredoxin-like
XX
SF   Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain
XX
FA   Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain
XX
DO   Aspartate carbamoyltransferase
XX
OS   Escherichia coli
XX
NC   1
XX
CN   [1]
XX
CH   B CHAIN; 8 START; 100 END;
//
ID   D1CS4A_
XX
EN   1CS4
XX
TY   SCOP
XX
SI   53931 CL; 54861 FO; 55073 SF; 55074 FA; 55077 DO; 55078 SO; 39418 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Ferredoxin-like
XX
SF   Adenylyl and guanylyl cyclase catalytic domain
XX
FA   Adenylyl and guanylyl cyclase catalytic domain
XX
DO   Adenylyl cyclase VC1, domain C1a
XX
OS   Dog (Canis familiaris)
XX
NC   1
XX
CN   [1]
XX


<font color=red>  [Part of this file has been deleted for brevity]</font>

XX
EN   1FX2
XX
TY   SCOP
XX
SI   53931 CL; 54861 FO; 55073 SF; 55074 FA; 55081 DO; 55082 SO; 39430 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Ferredoxin-like
XX
SF   Adenylyl and guanylyl cyclase catalytic domain
XX
FA   Adenylyl and guanylyl cyclase catalytic domain
XX
DO   Receptor-type monomeric adenylyl cyclase
XX
OS   Trypanosome (Trypanosoma brucei), different isoform
XX
NC   1
XX
CN   [1]
XX
CH   A CHAIN; . START; . END;
//
ID   D4AT1D1
XX
EN   4AT1
XX
TY   SCOP
XX
SI   53931 CL; 54861 FO; 54893 SF; 54894 FA; 54895 DO; 54896 SO; 39020 DD;
XX
CL   Alpha and beta proteins (a+b)
XX
FO   Ferredoxin-like
XX
SF   Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain
XX
FA   Aspartate carbamoyltransferase, Regulatory-chain, N-terminal domain
XX
DO   Aspartate carbamoyltransferase
XX
OS   Escherichia coli
XX
NC   1
XX
CN   [1]
XX
CH   D CHAIN; 8 START; 100 END;
//
</pre>
</td></tr></table><p>


<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>
DOMAINREP does not use any data files but does make use of the stamp 
"pdb.directories" file (see 'Notes').






<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>
<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 
<pre>
Reorder DCF file to identify representative structures.
Version: EMBOSS:6.2.0

   Standard (Mandatory) qualifiers:
  [-dcfinfile]         infile     This option specifies the name of DCF file
                                  (domain classification file) (input). A
                                  'domain classification file' contains
                                  classification and other data for domains
                                  from SCOP or CATH, in DCF format
                                  (EMBL-like). The files are generated by
                                  using SCOPPARSE and CATHPARSE. Domain
                                  sequence information can be added to the
                                  file by using DOMAINSEQS.
   -node               menu       [1] This option specifies the node fore
                                  reordering the file. Reordering can be done
                                  for any specified node in the SCOP or CATH
                                  hierarchies. For example by selecting
                                  'Class' entries belonging to the same Class
                                  will be reordered in turn. (Values: 1 (Class
                                  (SCOP)); 2 (Fold (SCOP)); 3 (Superfamily
                                  (SCOP)); 4 (Family (SCOP)); 5 (Class
                                  (CATH)); 6 (Architecture (CATH)); 7
                                  (Topology (CATH)); 8 (Homologous Superfamily
                                  (CATH)); 9 (Family (CATH)))
  [-dcfoutfile]        outfile    [test.scop] This option specifies the name
                                  of DCF file (domain classification file)
                                  (output). A 'domain classification file'
                                  contains classification and other data for
                                  domains from SCOP or CATH, in DCF format
                                  (EMBL-like). The files are generated by
                                  using SCOPPARSE and CATHPARSE. Domain
                                  sequence information can be added to the
                                  file by using DOMAINSEQS.

   Additional (Optional) qualifiers: (none)
   Advanced (Unprompted) qualifiers: (none)
   Associated qualifiers:

   "-dcfoutfile" associated qualifiers
   -odirectory2        string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</pre>
<P>
<table border cellspacing=0 cellpadding=3 bgcolor="#ccccff">
<tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-dcfinfile]<br>(Parameter 1)</td>
<td>infile</td>
<td>This option specifies the name of DCF file (domain classification file) (input). A 'domain classification file' contains classification and other data for domains from SCOP or CATH, in DCF format (EMBL-like). The files are generated by using SCOPPARSE and CATHPARSE. Domain sequence information can be added to the file by using DOMAINSEQS.</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-node</td>
<td>list</td>
<td>This option specifies the node fore reordering the file. Reordering can be done for any specified node in the SCOP or CATH hierarchies. For example by selecting 'Class' entries belonging to the same Class will be reordered in turn.</td>
<td><table><tr><td>1</td> <td><i>(Class (SCOP))</i></td></tr><tr><td>2</td> <td><i>(Fold (SCOP))</i></td></tr><tr><td>3</td> <td><i>(Superfamily (SCOP))</i></td></tr><tr><td>4</td> <td><i>(Family (SCOP))</i></td></tr><tr><td>5</td> <td><i>(Class (CATH))</i></td></tr><tr><td>6</td> <td><i>(Architecture (CATH))</i></td></tr><tr><td>7</td> <td><i>(Topology (CATH))</i></td></tr><tr><td>8</td> <td><i>(Homologous Superfamily (CATH))</i></td></tr><tr><td>9</td> <td><i>(Family (CATH))</i></td></tr></table></td>
<td>1</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-dcfoutfile]<br>(Parameter 2)</td>
<td>outfile</td>
<td>This option specifies the name of DCF file (domain classification file) (output). A 'domain classification file' contains classification and other data for domains from SCOP or CATH, in DCF format (EMBL-like). The files are generated by using SCOPPARSE and CATHPARSE. Domain sequence information can be added to the file by using DOMAINSEQS.</td>
<td>Output file</td>
<td>test.scop</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Additional (Optional) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Advanced (Unprompted) qualifiers</th>
</tr>

<tr>
<td colspan=5>(none)</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>"-dcfoutfile" associated outfile qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -odirectory2<br>-odirectory_dcfoutfile</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th align="left" colspan=5>General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</table>

<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of DOMAINREP is shown below.

<b>Here is a sample session with domainrep</b>
<p>

<p>
<table width="90%"><tr><td bgcolor="#CCFFFF"><pre>

% <b>domainrep </b>
Reorder DCF file to identify representative structures.
Domain classification file: <b>all.scop2</b>
Node at which to reorder
         1 : Class (SCOP)
         2 : Fold (SCOP)
         3 : Superfamily (SCOP)
         4 : Family (SCOP)
         5 : Class (CATH)
         6 : Architecture (CATH)
         7 : Topology (CATH)
         8 : Homologous Superfamily (CATH)
         9 : Family (CATH)
Select number. [1]: <b>2</b>
Domain classification output file [test.scop]: <b>all_rep1st.scop</b>

stamp -l ./domainrep-1234567890.1234.dom -s -n 2 -slide 5 -prefix ./domainrep-1234567890.1234 -d ./domainrep-1234567890.1234.set > ./domainrep-1234567890.1234.out

stamp -l ./domainrep-1234567890.1234.dom -s -n 2 -slide 5 -prefix ./domainrep-1234567890.1234 -d ./domainrep-1234567890.1234.set > ./domainrep-1234567890.1234.out

stamp -l ./domainrep-1234567890.1234.dom -s -n 2 -slide 5 -prefix ./domainrep-1234567890.1234 -d ./domainrep-1234567890.1234.set > ./domainrep-1234567890.1234.out

stamp -l ./domainrep-1234567890.1234.dom -s -n 2 -slide 5 -prefix ./domainrep-1234567890.1234 -d ./domainrep-1234567890.1234.set > ./domainrep-1234567890.1234.out

stamp -l ./domainrep-1234567890.1234.dom -s -n 2 -slide 5 -prefix ./domainrep-1234567890.1234 -d ./domainrep-1234567890.1234.set > ./domainrep-1234567890.1234.out

stamp -l ./domainrep-1234567890.1234.dom -s -n 2 -slide 5 -prefix ./domainrep-1234567890.1234 -d ./domainrep-1234567890.1234.set > ./domainrep-1234567890.1234.out


</pre></td></tr></table><p>
<p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a><p><p>

<!-- The DCF file /test_data/all.scop was read and the DCF file /test_data/domainrep/all_rep1st.scop was written in which the order of domains for each <i>family</i> was changed if necessary so that the representative structure of each family is given first.  -->

<br> 
<br>The following command line would achieve the same result.
<br>
<table><td bgcolor="#FFCCFF">
<pre> 
domainrep test_data/all.scop test_data/domainrep/all_rep1st.scop -node 4
</pre> 
</table>

<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
None.





<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>
<b>stamp modification</b>
<br>DOMAINREP will only run with with a version of stamp which has been modified
so that PDB id codes of length greater than 4 characters are acceptable.
This involves a trivial change to the stamp module getdomain.c (around line
number 155), a 4 must be changed to a 7 as follows:
<br><br>
<table><td bgcolor="#FFCCFF">
<pre>
temp=getfile(domain[0].id,dirfile,4,OUTPUT); 
<font color=red>< change to ... ></font>
temp=getfile(domain[0].id,dirfile,7,OUTPUT); 
</pre>
</table>

<br>
<b>stamp configuration</b>
<br>
stamp searches for pdb files with a certain prefix, extension and path as 
specified in the stamp "pdb.directories" file, which should look like 
something like this (change the text according to your system setup):

<br><br>
<table><td bgcolor="#FFCCFF">
<pre>
/data/pdb - -
/data/pdb _ .ent
/data/pdb _ .pdb
/data/pdb pdb .ent
/data/pdbscop _ _
/data/pdbscop _ .ent
/data/pdbscop _ .pdb
/data/pdbscop pdb .ent
./ _ _
./ _ .ent
./ _ .ent.z
./ _ .ent.gz
./ _ .pdb
./ _ .pdb.Z
./ _ .pdb.gz
./ pdb .ent
./ pdb .ent.Z
./ pdb .ent.gz
/data/CASS1/pdb/coords/ _ .pdb
/data/CASS1/pdb/coords/ _ .pdb.Z
/data/CASS1/pdb/coords/ _ .pdb.gz
</pre>
</table>

<br>



None




<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>

<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b> Domain classification file (for SCOP)</b></td>
<td> DCF format (EMBL-like format for domain classification data). </td>
<td> Classification and other data for domains from SCOP.  </td>
<td> <a href="scopparse.html">SCOPPARSE</a> </td>
<td> Domain sequence information can be added to the file by using DOMAINSEQS.  </td>
</tr>
<tr>
<td><b> Domain classification file (for CATH)</b></td>
<td> DCF format (EMBL-like format for domain classification data). </td>
<td> Classification and other data for domains from CATH.  </td>
<td> <a href="cathparse.html">CATHPARSE</a> </td>
<td> Domain sequence information can be added to the file by using DOMAINSEQS.  </td>
</tr>
<tr>
<td><b>Domain PDB file </b></td>
<td> PDB format for domain coordinate data. </td>
<td> Coordinate data for a single domain from SCOP or CATH. </td>
<td> <a href="domainer.html">DOMAINER</a> </td>
<td> N.A. </td>
</tr>
</table>






<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>
The protein structural alignment program stamp aligns multiple structures 
to a seed structure.  The best alignments will result when the seed is the
best available representative of the entire set, i.e. in 
structural terms, is more similar to the other structures than any other 
structure.  
<br>
<a href="domainalign.html">DOMAINALIGN</a> calls stamp to generate alignments
of domains belonging to the same node (e.g. family) in a DCF file.  DOMAINALIGN will produce better
alignments if the DCF file is reordered so that the 
representative structure of each node is given first, because the first
strucutres are used as the seeds.  The reordering of the DCF file is done by using DOMAINREP.






<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>

More information on stamp can be found at
http://www.compbio.dundee.ac.uk/manuals/stamp.4.2/




<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
     (Take this from "Sister applications" in the old documentation)
-->

<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<h2><a name="See also">See also</a></h2>
<table border cellpadding=4 bgcolor="#FFFFF0">
<tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="/embassy/structure/contacts.html">contacts</a></td>
<td>Generate intra-chain CON files from CCF files</td>
</tr>

<tr>
<td><a href="domainalign.html">domainalign</a></td>
<td>Generate alignments (DAF file) for nodes in a DCF file</td>
</tr>

<tr>
<td><a href="/embassy/domainatrix/domainreso.html">domainreso</a></td>
<td>Remove low resolution domains from a DCF file</td>
</tr>

<tr>
<td><a href="/embassy/structure/interface.html">interface</a></td>
<td>Generate inter-chain CON files from CCF files</td>
</tr>

<tr>
<td><a href="/embassy/signature/libgen.html">libgen</a></td>
<td>Generate discriminating elements from alignments</td>
</tr>

<tr>
<td><a href="/embassy/signature/matgen3d.html">matgen3d</a></td>
<td>Generate a 3D-1D scoring matrix from CCF files</td>
</tr>

<tr>
<td><a href="/emboss/apps/psiphi.html">psiphi</a></td>
<td>Calculates phi and psi torsion angles from protein coordinates</td>
</tr>

<tr>
<td><a href="/embassy/signature/rocon.html">rocon</a></td>
<td>Generates a hits file from comparing two DHF files</td>
</tr>

<tr>
<td><a href="/embassy/signature/rocplot.html">rocplot</a></td>
<td>Performs ROC analysis on hits files</td>
</tr>

<tr>
<td><a href="seqalign.html">seqalign</a></td>
<td>Extend alignments (DAF file) with sequences (DHF file)</td>
</tr>

<tr>
<td><a href="/embassy/domsearch/seqfraggle.html">seqfraggle</a></td>
<td>Removes fragment sequences from DHF files</td>
</tr>

<tr>
<td><a href="/embassy/domsearch/seqsearch.html">seqsearch</a></td>
<td>Generate PSI-BLAST hits (DHF file) from a DAF file</td>
</tr>

<tr>
<td><a href="/embassy/domsearch/seqsort.html">seqsort</a></td>
<td>Remove ambiguous classified sequences from DHF files</td>
</tr>

<tr>
<td><a href="/embassy/domsearch/seqwords.html">seqwords</a></td>
<td>Generates DHF files from keyword search of UniProt</td>
</tr>

<tr>
<td><a href="/embassy/signature/siggen.html">siggen</a></td>
<td>Generates a sparse protein signature from an alignment</td>
</tr>

<tr>
<td><a href="/embassy/signature/siggenlig.html">siggenlig</a></td>
<td>Generates ligand-binding signatures from a CON file</td>
</tr>

<tr>
<td><a href="/embassy/signature/sigscan.html">sigscan</a></td>
<td>Generates hits (DHF file) from a signature search</td>
</tr>

<tr>
<td><a href="/embassy/signature/sigscanlig.html">sigscanlig</a></td>
<td>Searches ligand-signature library & writes hits (LHF file)</td>
</tr>

</table>


<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
None.






<!-- AUTHORS -->
<a name="13.0"><br><br><br><H2> 13.0  AUTHORS                   </H2></a>

Mat Blades <a href="mailto:Matthew.Blades@astrazeneca.com">(Matthew.Blades@astrazeneca.com)</a>
<br><br>
Jon Ison <a href="mailto:jison@ebi.ac.uk">(jison@ebi.ac.uk)</a>
<br>
The European Bioinformatics Institute 
Wellcome Trust Genome Campus 
Cambridge CB10 1SD 
UK 




<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

Please cite the authors and EMBOSS.
<br><br>
<i>Rice P, Longden I and Bleasby A (2000) "EMBOSS - The European
Molecular Biology Open Software Suite"  Trends in Genetics,
15:276-278.

<p>
See also <a href="http://emboss.sourceforge.net/">http://emboss.sourceforge.net/</a></i>


<H3>14.1 Other useful references  </H3>

<br>

</BODY>
</HTML>
