<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: SEQNR documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">
<!--#include virtual="/apps/header1.inc" -->
<H2> SEQNR documentation
<!--#include virtual="/apps/header2.inc" -->
<!-- END OF HEADER -->







<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<br><H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>

<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>


<!--#include file="inc/seqnr.ione" -->




<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
SEQNR removes redundancy from DHF files (domain hits files) or other files of sequences.  A directory of DHF files (all sequences) is read and a directory of new DHF files (non-redundant sequences) plus (optionally) a second directory of DHF files (redundant sequences) is written.  Optionally, up to two further directories of filter sequences may be read: these are considered in the redundancy calculation but never appear in the output files.   Typically, one of the further directories contains DHF files each with a single sequence and the other DAF files (domain alignment files) each containing a sequence alignment, but any sequence(s) may be given.  Each filter directory must contain a file for each file in main input directory and the files must have the same base name.  For example, sequences from "family.dhf" and "family.daf" are considered for the input DHF file "family.hits". 
<br>Redundancy is removed at either (i) a user-defined threshold of sequence similarity or (ii) a user-defined range of threshold sequence similarity. Files of sequences in any supported format may be read and written (not just DHF or DAF files). A log file is also written.
<br>The path for all files (input and output) are specified by the user.  The file extensions are set in the ACD file.  The name of the log file is set by the user.





<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.

     Cannot use the test data files because they might be empty or need 
     hand-editing 

     Use "<b>SEQNR</b> reads any normal sequence USAs." if
     appropriate.
-->
<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>
The format of the domain hits file is described in <a href="seqsearch.html">SEQSEARCH documentation.</a>
<br>The format of the domain alignment file is described in <a href="domainalign.html">DOMAINALIGN documentation</a>. 
<br>If other sequences or sequence sets (aligned or unaligned) are used as input, all of the common file formats are supported.

<!--#include file="inc/seqnr.input" -->


<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>
The format of the domain hits file is described in <a href="seqsearch.html">SEQSEARCH documentation.</a>


<!--#include file="inc/seqnr.output" -->

<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>
SEQNR requires a residue substitution matrix.



<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>
<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 
<pre>
<!--#include file="inc/seqnr.ihelp" -->
</pre>
</td></tr></table>
<P>
<!--#include file="inc/seqnr.itable" -->


<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of SEQNR is shown below.

<!--#include file="inc/seqnr.usage" -->

<!-- Two domains within the same family are classified as redundant if they share at least 40% sequence similarity; the default residue substitution matrix EBLOSUM62 and default gap insertion and extension penalties were used for the alignments. -->


<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
None.





<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>
None.



<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b>Domain hits file </b></td>
<td> DHF format (FASTA-like). </td>
<td> Database hits (sequences) with domain classification information.  The hits are relatives to a SCOP or CATH family (or other node in the structural hierarchies) and are found from a search of a discriminating element (e.g. a protein signature, hidden Markov model, simple frequency matrix, Gribskov profile or Hennikoff profile) against a sequence database. </td>
<td> <a href="seqsearch.html">SEQSEARCH</a> (hits retrieved by PSIBLAST). <a href="sigscan.html">SIGSCAN</a> (hits retrieved by sparse protein signature). <a href="libscan.html">LIBSCAN</a> (hits retrieved by various types of HMM and profile). </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>Domain alignment file </b></td>
<td> DAF format (CLUSTAL-like format with domain classification information). </td>
<td> Contains a sequence alignment of domains belonging to the same SCOP or CATH family.  The file is annotated with domain family classification information.</td>
<td> <a href="domainalign.html">DOMAINALIGN</a> (structure-based sequence alignment of domains of known structure). </td>
<td> DOMAINALIGN alignments can be extended with sequence relatives (of unknown structure) to the family in question by using SEQALIGN. </td>
</tr>
</table>




<!--#include file="inc/seqnr.history" -->
<!--#include file="inc/seqnr.comment" -->




<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>

Redundancy in a database or other collection of sequences occurs when one or more similar sequences are present. The inclusion of very similar sequences in certain analyses will introduce undesirable bias. For example, a family may possess 100 sequences in the sequence database, but 90 of these might be essentially the same sequence, e.g. very close relatives or mutations of a single sequence.  Although 100 sequences are known, the family only contains 11 sequences that are essentially unique.  For many applications it is desirable or even essential to remove redundant sequences from a set in order to produce a smaller set that is representative of the whole. SEQNR removes redundancy from an input file of sequences, either at a single threshold of sequence similiarty (e.g. 40%) or within a threshold range of sequence similiarty (e.g. 40% - 70%).





<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>

Redundancy is calculated for each DHF file in the input directory in turn.  The procedure is as follows. 1.  Create a list of sequences from the main input directory.  2. Add sequences from both filter directories (if specified) to list but mark them up (they are considered in the redundancy calculation but never appear in the output files).  3. Identify redundant domains. 4. Write non-redundant domains to main output directory.  5.  If specified, write redundant domains to output directory.





<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
     (Take this from "Sister applications" in the old documentation)
-->
<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>


<!--#include file="inc/seqnr.isee" -->

<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
The log file might contain the following messages:
<br><br><i>embHitlistReadFasta call failed in seqnr</i> (By default SEQNR expects a domain hits file in the main input directory and the first (singlets) filter directory.  This message is given if some other type of sequence file is given.  Its not really an error, just a warning).
<br><br><i>Empty input file my.file</i> (This is given if the main input file 'my.file' did not contain any sequences.  No output file is generated for this file.)
<br><br><i>Empty singlets filter file my.file</i> (This is given if the singlets filter file 'my.file' did not contain any sequences.  An output file is still generated for this file.)
<br><br><i>Empty sets filter file my.file</i> (This is given if the sets filter file 'my.file' did not contain any sequences.  An output file is still generated for this file.)
<br><br><i>ajDmxScopalgRead call failed in seqnr my.file</i> (By default SEQNR expects a domain alignment file in the second (sets) filter directory.  This message is given if some other type of sequence sets file (called 'my.file') is given.  Its not really an error, just a warning).


<!-- AUTHORS -->
<a name="13.0"><br><br><br><H2> 13.0  AUTHORS                   </H2></a>
Jon Ison <a href="mailto:jison@ebi.ac.uk">(jison@ebi.ac.uk)</a>
<br>
The European Bioinformatics Institute 
Wellcome Trust Genome Campus 
Cambridge CB10 1SD 
UK

<br><br><i>based on an original program written by</i><br><br>
Ranjeeva Ranasinghe





<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

<!--#include file="inc/embossref.ihtml" -->

<H3>14.1 Other useful references  </H3>

<br>

</BODY>
</HTML>
