<!-- START OF HEADER -->
<HTML><HEAD>
<TITLE> EMBASSY: SEQSEARCH documentation. </TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" text="#000000">

<!--#include virtual="/apps/header1.inc" -->
<H2> SEQSEARCH documentation
<!--#include virtual="/apps/header2.inc" -->
<!-- END OF HEADER -->






<!-- CONTENTS
     This always includes the sections below.
     Other subsections can be added for individual applications.
-->
<br><H2>CONTENTS </H2>
<b> <a href="#1.0">1.0     SUMMARY                   </a></b><br>
<b> <a href="#2.0">2.0     INPUTS & OUTPUTS          </a></b><br>
<b> <a href="#3.0">3.0     INPUT FILE FORMAT         </a></b><br>
<b> <a href="#4.0">4.0     OUTPUT FILE FORMAT        </a></b><br>
<b> <a href="#5.0">5.0     DATA FILES                </a></b><br>
<b> <a href="#6.0">6.0     USAGE                     </a></b><br>    
<b> <a href="#7.0">7.0     KNOWN BUGS & WARNINGS     </a></b><br>    
<b> <a href="#8.0">8.0     NOTES                     </a></b><br>
<b> <a href="#9.0">9.0     DESCRIPTION               </a></b><br>
<b> <a href="#10.0">10.0   ALGORITHM                 </a></b><br>
<b> <a href="#11.0">11.0   RELATED APPLICATIONS      </a></b><br>
<b> <a href="#12.0">12.0   DIAGNOSTIC ERROR MESSAGES </a></b><br>
<b> <a href="#13.0">13.0   AUTHORS                   </a></b><br>
<b> <a href="#14.0">14.0   REFERENCES                </a></b><br>

<!-- SUMMARY
     Succint description of the application, particularly its inputs, outputs
     and what it does.  The same text is given at the top of the source (.c)
     file and in the <documentation> attribute of the <application definition>
     of the ACD file.
-->
<a name="1.0"></a>
<br><br><br><H2> 1.0   SUMMARY  </H2>

<!--#include file="inc/seqsearch.ione" -->


<!-- INPUTS & OUTPUTS
     Short summary of the application inputs and outputs in its different 
     modes of usage (if appropriate). More detail than the summary.
-->
<a name="2.0"></a>
<br><br><br><H2> 2.0   INPUTS & OUTPUTS          </H2>
SEQSEARCH reads a directory of i. single protein sequences or ii. set of protein sequences (aligned or unaligned) and generates a DHF file ('domain hits file') of sequence relatives (hits) for each file in the input directory.  The hits are sequence relatives to the input sequences and are found by using PSIBLAST. Only unique hits are generated; only one of a set of identical hits returned by PSIBLAST is retained.
<br><br>Typically, aligned sequences within a DAF file ('domain alignment file') are input and the DHF file output is annotated with domain classification data.
<br><br>PSIBLAST must be installed on the system that is running SEQSEARCH (see 'Notes' below). The base name of an input file is used as the base name for the corresponding output file. The paths and extensions for the sequence files (input) and domain hits files (output) are specified by the user.  The name of the BLAST-indexed database to search are also user-specified.  A log file is also written.




<!-- INPUT FILE FORMAT 
     Description and example(s) of input file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.

     Cannot use the test data files because they might be empty or need 
     hand-editing 

     Use "<b>SEQSEARCH</b> reads any normal sequence USAs." if
     appropriate.
-->
<a name="3.0"></a>
<br><br><br><H2> 3.0   INPUT FILE FORMAT         </H2>

The format of the domain alignment file is described in <a href="domainalign.html">DOMAINALIGN documentation</a>. 
<br>If other sequences or sequence sets (aligned or unaligned) are used as input, all of the common file formats are supported.

<!--#include file="inc/seqsearch.input" -->

<!-- OUTPUT FILE FORMAT 
     Description and example(s) of output file formats.  Should provide enough
     information to write and parse the file.  Should describe the format in 
     unusual cases - null input, etc.  
     If the standard description of the avalable report formats is required, 
     use:   #include file="inc/reportformats.ihtml"
     Use "Outputs a graph to the specified graphics device."
     or  "outputs a report format file. The default format is ..."
     if appropriate.
-->
<a name="4.0"></a>
<br><br><br><H2> 4.0   OUTPUT FILE FORMAT         </H2>
SEQSEARCH generates a domain hits file in FASTA-like format (<a href="#Figure1">Figure 1</a>).

<a name="Figure1"></a>
<br><br> <b>Figure 1 DHF file (FASTA-like format)</b>
<br>The file (Figure 1) contains two lines per hit. The first contains a description of the hit in 16 text tokens delimited by '^'.  The tokens are as follows (a '.' is given where a token does not have a value).

The first 4 tokens are specific to the sequence of the hit:
<UL>
<LI>(i) Accession number of the hit.  
<LI>(ii)  Database code from Uniprot.  
<LI>(iii - iv) Start and end positions of the hit relative to the full length sequence in the uniprot database (files of these type may also be generated by using SEQWORDS in which case a '.' will be given for these records - see <a href="seqwords.html">SEQWORDS documentation </a>). 
</UL>

The next 9 tokens are specific to the domain (or domain family or other node) for which the hit was generated:
<UL>
<LI>(v)  Type of domain; currently either 'SCOP' or 'CATH' is given.
<LI>(vi)  SCOP or CATH domain identifier.  This is a 7-character code that uniquely identifies the domain in SCOP or CATH.
<LI>(vii) SCOP or CATH node unique identifier. For example, if the domain alignment file was for a SCOP family, the SCOP Sunid for the family would be given. This number uniquely identifies the node (i.e. family in this case) in the raw SCOP or CATH parsable files. 
<LI>(viii) Domain class.  Textual description of the 'Class' (SCOP and CATH domains). 
<LI>(ix) Domain architecture. Textual description of the 'Architecture' (CATH only). 
<LI>(x) Domain topology. Textual description of the 'Topology' (CATH only).  
<LI>(xi) Domain fold. Textual description of the 'Fold' (SCOP domains only).  
<LI>(xii) Domain superfamily. Textual description of the 'Superfamily' (SCOP and CATH domains).  
<LI>(xiii) Domain family. Textual description of the 'Fold' (SCOP only). 
</UL>

The next 4 tokens are specific to the hit itself:
<UL>
<LI>(xiv) Model type.  The type of model that was used to generate the hit. May have a value of "PSIBLAST" (from PSIBLAST), "HMMER" (hidden Markov model from the HMMER package), "SAM" (hidden Markov model from the SAM package), SPARSE (sparse protein signature), HENIKOFF (Henikoff profile) or GRIBSKOV (Gribskov profile). A value of "PSIBLAST" is written by SEQSEARCH, a value of "KEYWORD" is written by SEQWORDS.
<LI>(xv) SC - Score of hit. A floating point value that is the score from psiblast (or other search algorithm). 
<LI>(xvi) P-value of hit from search algorithm. 
<LI>(xvii) E-value of hit from search algorithm. The second line contains the protein sequence.
</UL>


<!--#include file="inc/seqsearch.output" -->


<!--
<br><br><b>Figure 1   Excerpt from a SEQSEARCH DHF file (FASTA-like format output) </b>
<table><td bgcolor="#CFCCFF">
<pre>
> Q9WVI4^.^513^667^.^55074^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Adenylyl and guanylyl cyclase catalytic domain^Adenylyl and guanylyl cyclase catalytic domain^PSIBLAST^113.00^0.000e+00^2.000e-25
RKFDDVTMLFSDIVGFTAICAQCTPMQVISMLNELYTRFDHQCGFLDIYKVETIGDAYCVASGLHRKSLCHAKPIALMALKMMELSEEVLTPDGRPIQMRIGIHSGSVLAGVVGVRMPRYCLFGNNVTLASKFESGSHPRRINISPTTYQLLKRE
> Q9ERL9^.^475^626^.^55074^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Adenylyl and guanylyl cyclase catalytic domain^Adenylyl and guanylyl cyclase catalytic domain^PSIBLAST^101.00^0.000e+00^9.000e-22
KKFSEVTMLFSDIVGFTAICSQCSPLQVITMLNALYTRFDQQCGELDVYKVETIGDAYCVAGGLHRESDTHAVQIALMALKMMELSNEVMSPHGEPIKMRIGLHSGSVFAGVVGVKMPRYCLFGNNVTLANKFESCSVPRKINVSPTTYRLL
> Q9DGG6^.^1032^1200^.^55074^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Adenylyl and guanylyl cyclase catalytic domain^Adenylyl and guanylyl cyclase catalytic domain^PSIBLAST^72.70^0.000e+00^4.000e-13
YSKNHDSGGVIFASIVNFSEFYEENYEGGKECYRVLNELIGDFDELLSKPHYSSIEKIKTIGATYMAASGLNTSQCQDSNHPHGHLQTLFEFAKEMMRVVDDFNNNMLWFNFKLRIGFNHGPLTAGVIGTTKLLYDIWGDTVNIASRMDTIGVECRIQVSEETYRILSK
> Q9DGG6^.^369^560^.^55074^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Adenylyl and guanylyl cyclase catalytic domain^Adenylyl and guanylyl cyclase catalytic domain^PSIBLAST^172.00^0.000e+00^5.003e-43
IIFRPFKMQQIEQVSILFADIVGFTKMSANKSAHALVGLLNDLFGRFDRLCEDTKCEKISTLGDCYYCVAGCPEPRADHAYCCIEMGLGMIKAIEQFCQEKKEMVNMRVGVHTGTVLCGILGMRRFKFDVWSNDVNLANLMEQLGVAGKVHISEATAKYLDDRYEMEDGKVTERVGQSAVADQLKGLKTYLI
> Q99396^.^250^473^.^55074^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Adenylyl and guanylyl cyclase catalytic domain^Adenylyl and guanylyl cyclase catalytic domain^PSIBLAST^182.00^0.000e+00^0.000e+00
IMRNKRDNDNAPKELADPVTLIFTDIESSTAQWATQPELMPDAVATHHSMVRSLIENYDCYEVKTVGDSFMIACKSPFAAVQLAQELQLRFLRLDWGTTVFDEFYREFEERHAEEGDGKYKPPTARLDPEVYRQLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGQTANTAARTESVGNGGQVLMTCETYHSLSTAERSQFDVTPLGGVPLRGVSEPVEVYQLN
> Q99280^.^877^1092^.^55074^Alpha and beta proteins (a+b)^.^.^Ferredoxin-like^Adenylyl and guanylyl cyclase catalytic domain^Adenylyl and guanylyl cyclase catalytic domain^PSIBLAST^204.00^0.000e+00^0.000e+00
NDSAPKEPTGPVTLIFTDIESSTALWAAHPDLMPDAVATHHRLIRSLITRYECYEVKTVGDSFMIASKSPFAAVQLAQELQLRFLRLDWETNALDESYREFEEQRAEGECEYTPPTAHMDPEVYSRLWNGLRVRVGIHTGLCDIRYDEVTKGYDYYGRTSNMAARTESVANGGQVLMTHAAYMSLSGEDRNQLDVTTLGATVLRGVPEPVRMYQLN
</pre>
</table>


-->

<!-- DATA FILES         
     Any data files used (e.g. translation table file, substitution matrix 
     etc.  This includes example data file formats if they are not obvious.
     For a standard description of what data files are and how embossdata can
     be used to inspect and retrieve them, use:
     #include file="inc/localfiles.ihtml" 
-->

<a name="5.0"></a>
<br><br><br><H2> 5.0   DATA FILES                </H2>

SEQSEARCH does not requires any data files.






<!-- USAGE
     Example usage, as run from the command-line.
     Many examples illustrating different behaviours is good.
-->
<a name="6.0"></a>
<br><br><br><H2> 6.0   USAGE                     </H2>
<H3> 6.1   COMMAND LINE ARGUMENTS </H3>
 
<pre>
<!--#include file="inc/seqsearch.ihelp" -->
</pre>
</td></tr></table>
<P>
<!--#include file="inc/seqsearch.itable" -->


<H3> 6.2 EXAMPLE SESSION </H3>
An example of interactive use of SEQSEARCH is shown below.

<!--#include file="inc/seqsearch.usage" -->

<!-- All domain alignment files (with the file extension of .daf specified in the ACD file) were read from the directory /test_data; in this case two domain alignment files 54894.salign and 55074.salign were read. Sets of sequences extracted from these files were used to search the sequence database swissprot by using psiblast.  psiblast was configured to perform 1 iteration with a threshold E-value for acceptance of a hit of 0.0001 and no more than 100 hits were generated from each iteration.  Domain hits files were written to /test_data/seqsearch ( the file extension .dhf was specified in the ACD file); in this case two files /test_data/54894.dhf and /test_data/55074.dhf were written.  A log file called /test_data/seqsearch/seqsearch.log was also written. -->




<!-- KNOWN BUGS & WARNINGS
     Bugs that have not yet been fixed, easily missued features, problems
     and caveats etc. Potentially stupid things the program will let you do.
-->
<a name="7.0"></a>
<br><br><br><H2> 7.0   KNOWN BUGS & WARNINGS     </H2>
None.




<!-- NOTES
     Important general remarks, including:
       Restrictions.
       Interesting behaviour.
       Useful things you can do with this program.
       Future plans.
       etc.
-->
<a name="8.0"></a>
<br><br><br><H2> 8.0   NOTES                     </H2>
<b>  1. Use of psiblast</b><br>
psiblast must be installed on the system that is running SEQSEARCH.

<br><br>When running SEQSEARCH at the HGMP it is essential that the command 'use blast_v2' (which runs the script /packages/menu/USE/blast_v2) is given before it is run. 

<br><br>SEQSEARCH requires a blast-indexed database to be present, i.e. both the sequence and index file must be present on the system.  The name of the database to search specified in the acd file is that which is given as the -d parameter to blastpgp (e.g. blastpgp -d swissprot).



<H3> 8.1   GLOSSARY OF FILE TYPES    </H3>
<table BORDER CELLSPACING=5 CELLPADDING=5 BGCOLOR="#f5f5ff" >
<tr>
<td><b>FILE TYPE</b></td>
<td><b>FORMAT</b></td>
<td><b>DESCRIPTION</b></td>
<td><b>CREATED BY <b></td>
<td><b>SEE ALSO</b></td>
</tr>
<tr>
<td><b>Domain hits file </b></td>
<td> DHF format (FASTA-like). </td>
<td> Database hits (sequences) with domain classification information.  The hits are relatives to a SCOP or CATH family (or other node in the structural hierarchies) and are found from a search of a discriminating element (e.g. a protein signature, hidden Markov model, simple frequency matrix, Gribskov profile or Hennikoff profile) against a sequence database. </td>
<td> <a href="seqsearch.html">SEQSEARCH</a> (hits retrieved by PSIBLAST). <a href="sigscan.html">SIGSCAN</a> (hits retrieved by sparse protein signature). <a href="libscan.html">LIBSCAN</a> (hits retrieved by various types of HMM and profile). </td>
<td> N.A. </td>
</tr>
<tr>
<td><b>Domain alignment file </b></td>
<td> DAF format (CLUSTAL-like format with domain classification information).  </td>
<td> Contains a sequence alignment of domains belonging to the same SCOP or CATH family.  The file is annotated with domain family classification information.</td>
<td> <a href="domainalign.html">DOMAINALIGN</a> (structure-based sequence alignment of domains of known structure). </td>
<td> DOMAINALIGN alignments can be extended with sequence relatives (of unknown structure) to the family in question by using SEQALIGN. </td>
</tr>
</table>



<!--#include file="inc/seqsearch.history" -->
<!--#include file="inc/seqsearch.comment" -->



<!-- DESCRIPTION
     A complete, non-technical, user-level description of the application.
-->
<a name="9.0"></a>
<br><br><br><H2> 9.0   DESCRIPTION               </H2>
By using homology search tools such as blast it is possible to find relatives to a group of related proteins (family, superfamily etc), given one or more sequences belonging to the group of interest.  For example, when using psiblast it is possible to use a sequence alignment as the seed with which to search a sequence database.  Performing such searches for large datasets such as all families within SCOP or CATH potentially requires a lot of time for preparation of datasets, running jobs and so on, in addition to the compute time required for the searches themselves.  SEQSEARCH automatically performs a psiblast search of a sequence database for each file in a directory of sequences or sets of sequences.  These sequences are used for the searches.  Typically, the directory contains DAF files (domain alignment files) and the alignments are for a certain node (e.g. family, superfamily etc) from SCOP or CATH. 



<!-- ALGORITHM
     A technical description of algorithmic aspects, describing exactly how
     the key aspects of the application work.
-->
<a name="10.0"></a>
<br><br><br><H2> 10.0  ALGORITHM                 </H2>
None.


<!-- RELATED APPLICATIONS
     Other applications that either generate the input, use the output or 
     are in some other way related to the application are described here. 
     (Take this from "Sister applications" in the old documentation)
-->
<a name="11.0"></a>
<br><br><br><H2> 11.0  RELATED APPLICATIONS      </H2>

<!--#include file="inc/seqsearch.isee" -->


<!-- DIAGNOSTIC ERROR MESSAGES
     Description of error messages or log file, if one is written.
-->
<a name="12.0"></a>
<br><br><br><H2> 12.0  DIAGNOSTIC ERROR MESSAGES </H2>
The following 3 types of message might appear in the log file:
<br><br>WARN  Could not open for reading my.file
<br>WARN  No PSIBLAST hits therefore no output file my.file
<br>WARN  Could not open for writing my.file


<!-- AUTHORS -->
<a name="13.0"><br><br><br><H2> 13.0  AUTHORS                   </H2></a>
Ranjeeva Ranasinghe <a href="mailto:rranasin@rfcgr.mrc.ac.uk">(rranasin@rfcgr.mrc.ac.uk)</a>
<br><br>
Jon Ison <a href="mailto:jison@rfcgr.mrc.ac.uk">(jison@rfcgr.mrc.ac.uk)</a>
<br>MRC Rosalind Franklin Centre for Genomics Research, Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK



<!-- REFERENCES
     Quote the paper where the application was first published, described, used etc. 
-->
<a name="14.0"></a>
<br><br><br><H2> 14.0  REFERENCES                </H2>

<!--#include file="inc/embossref.ihtml" -->

<H3>14.1 Other useful references  </H3>
Altschul et al, Nuc. Acids. Res. 25:3389-3402, 1997
<br>

</BODY>
</HTML>
