/* @source sigscan application
**
** Scans a signature against swissprot and writes a signature hits file.
**
** @author: Copyright (C) Jon Ison (jison@hgmp.mrc.ac.uk)
** @@
**
** This program is free software; you can redistribute it and/or
** modify it under the terms of the GNU General Public License
** as published by the Free Software Foundation; either version 2
** of the License, or (at your option) any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
** 
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
**

*******************************************************************************
**  Application name
**  sigscan 
**  
**  
**  Summary
**  Scans a signature against swissprot and writes a signature hits file.
**  
**  
**  Input and Output
**  sigscan reads a protein signature file, scans the signature against a 
**  protein sequence database and generates a signature hits file (of scored 
**  hits to database sequences) and a signature alignment file (containing the
**  corresponding signature-sequence alignments).   A scop validation file can
**  be provided in which case a classification of hits is provided in the 
**  signature hits file.  The names of signature file, signature hits file, 
**  signature alignment file and validation files are provided by the user.
**  The user specifies a maximum number of false hits that are written to 
**  the signature hits files.
**  
**  Sister applications
**  A 'signature file' contains a sparse sequence signature.  The files are 
**  generated by siggen.
**  A 'scop validation file' contains sequence relatives (hits) for each of a
**  number of different scop families, superfamilies and folds.  The file 
**  contains the collated results from psiblast searches of a sequence database
**  for the indvidual scop families; hits of unambiguous family assignment are
**  listed under their respective family, otherwise a hit is assigned as 
**  relatives to a scop superfamily or fold instead.  The scop validation file
**  is generated by seqnr and is in embl-like format.
**  Signature hits and alignment files are used by sigplot.
**
**  
**  
**  Notes
**  Definition of classes of hit
**  A primary and secondary classification of hits are used and correspond to 
**  two different columns in the signature hits file (see 'Output file format'
**  below).
**  The primary classification is an objective definition of the hit and has 
**  one of the following values:
**  SEED - the sequence was included in the original alignment from which the 
**  signature was generated.
**  HIT - A protein which was detected by psiblast  (see psiblasts.c) to 
**  be a homologue to at least one of the proteins in the family from which 
**  the signature was derived. Such proteins are identified by the 'HIT' 
**  record in the scop families file.
**  OTHER - A true member of the family but not a homologue as detected by 
**  psi-blast. Such proteins may have been found from the literature and 
**  manually added to the scop families file or may have been detected by the 
**  EMBOSS program swissparse (see swissparse.c). They are identified in the 
**  SCOP families file by the 'OTHER' record.
**  CROSS - A protein which is homologous to a protein of the same fold,
**  but differnt family, of the proteins from which the signature was
**  derived.
**  FALSE - A homologue to a protein with a different fold to the family
**  of the signature.
**  UNKNOWN - The protein is not known to be CROSS, FALSE or a true hit (a 
**  SEED, HIT or OTHER).
**  The secondary classification is provided for convenience and a value as 
**  follows:
**  Hits of SEED, HIT and OTHER classification are all listed as TRUE.
**  Hits of CROSS, FALSE or UNKNOWN objective classification are listed as CROSS, 
**  FALSE or UNKNOWN respectively.
** 
**  The subjective column allows for hand-annotation of the hits files so that 
**  proteins of UNKNOWN objective classification can re-classified by a human 
**  expert as TRUE, FALSE, CROSS or otherwise left as UNKNOWN for the purpose of 
**  generating signature performance plots with the EMBOSS application sigplot.
**  
**  
**  p-values and E-values
**  sigscan does not generate p-values or E-values.  Signature hits files
**  containing p-values and E-values generated from scans of sparse protein 
**  signatures and also hidden Markov models and various other types of profile
**  may be generated by using dbscan.
**  
**  
**  Known bugs & warnings
**  sigscan presumes that SCOP family names are unique. If this were not the
**  case, changes to ajXyzClassifyHits would have to be made.
**  
**  In the case a signature file is generated by hand, it is essential that the 
**  gap data given is listed in order of increasing gap size.  
**  
**  
**  
**  Description
**  This program is part of a suite of EMBOSS applications that directly or 
**  indirectly make use of the protein structure databases pdb and scop.  
**  This program is part of an experimental analysis pipeline described in an
**  accompanying document.  We provide the software in the hope that it will
**  be useful.  The applications were designed for specific research purposes
**  and may not be useful or reliable in contexts other than the described 
**  pipeline.  The development of the suite was coordinated by Jon Ison to
**  whom enquiries and bug reports should be sent (email jison@hgmp.mrc.ac.uk).
**  
**  See Ison et al and Daniel et al for a description of protein signatures and
**  their application.
**  
**  
**  
**  Algorithm
**  The algorithm is based on approach first described by Daniel et al (1999) 
**  that was applied to the definition of protein families (Ison et al, 2000).
**  
**  
**  
**  Usage 
**  An example of interactive use of sigscan is shown below.
**  
**  Unix % sigscan
**  Scans a signature against swissprot and writes a signature hits file
**  Name of signature file (input): /test_data/54894.sig
**  Name of swissprot sequence database to search: /test_data/test.seq
**  Minimum length (residues) of overlap required for two hits with the same code to be counted as the same hit. [20]:
**  Residue substitution matrix [EBLOSUM62]: 
**  Name of validation (input): /test_data/scop.all
**  Gap insertion penalty [10]: 
**  Gap extension penalty [0.5]: 
**  N-terminal matching options
**           1 : Align anywhere and allow only complete signature-sequence fit
**           2 : Align anywhere and allow partial signature-sequence fit
**           3 : Use empirical gaps only
**  Select number [1]: 1
**  Number of hits to output [100]: 
**  Max. number of false hits to output [50]:
**  Name of signature hits file (output) [test.sighits]: /test_data/54894.sighits
**  Name of signature alignments file (output) [test.sigalign]: /test_data/54894.sigalign
**  Signature file read ok
**  Signature compiled ok
**  Targets read ok
**  Signature aligned to db ok
**  Hit classified ok
**  Hits file written ok
**  Alignments file written ok
**  Unix % 
**  
**  The signature file /test_data/54894.sig was scanned against the swissprot 
**  database /test_data/test.seq to generate a signature hits file called 
**  /test_data/54894.sighits and a signature alignment file called 
**  /test_data/54894.sigalign.  The output files contained no more than 50 
**  false hits or 100 hits in total.  Hits in the signature hits file were 
**  classified according to 
**  sequences whose classification has already been inferred and is given in
**  the validation file /test_data/scop.all.  In order for a hit in the 
**  signature hits file to be classed as the same as a hit in the validation
**  file, a shared region of at least 20 identical residues is required.  When
**  aligning and scoring the signature to a protein sequence, the residue 
**  substitution matrix EBLOSUM62, a gap insertion penalty of 10 and a gap 
**  extension penalty of 0.5 was used.  The first signature position was 
**  permitted to be aligned anywhere in the sequence without attracing a gap
**  penalty and only complete signature-sequence fits were allowed, i.e. every
**  signature position had to be aligned or otherwise the sequence was 
**  rejected (sequences smaller than the sum of the gap distances in the 
**  signature were rejected).
**  
**  The following command line would achieve the same result.
**  sigscan /test_data/54894.sig /test_data/54894.sighits 
**  /test_data/54894.sigalign -database /test_data/test.seq -targetf 
**  /test_data/scop.all -thresh 20 -sub EBLOSUM62 -gapo 10 -gape 0.5 -nterm 1
**   -nhits 100 -nhits 50
**  
** 
**  
**  Input file format
**  The format of the scop validation is exactly the same as the scop families
**  file, described in seqsearch.c.  The format of the signature file is 
**  described in siggen.c
** 
**   
**  
**  Output file format
**  Excerpts from a signature hits (below) are shown. The records used 
**  are are as follows
**  (1) DE - bibliographic information. The text 'Results of signature search'
**  is always given.
**  Four SCOP classification records are given:
**  (2)  CL - Domain class.  It is identical to the text given after 'Class' in 
**  the scop classification file (see documentation for the EMBOSS application 
**  scope).
**  (3)  FO - Domain fold.  It is identical to the text given after 'Fold' in 
**  the scop classification file (see scope documentation).
**  (4)  SF - Domain superfamily.  It is identical to the text given after 
**  'Superfamily' in the scop classification file (see scope documentation).
**  (5)  FA - Domain family. It is identical to the text given after 'Family' in 
**  the scop classification file (see scope documentation).
**  (6)  SI - SCOP Sunid's of the family. This number uniquely identifies the 
**  family in the SCOP parsable files.
**  (7) HI - hit data.  The data are as follows (column numbers are given in 
**  parentheses). (i) HI is always given. (ii) Rank-order of the hit. (iii) 
**  Accession number. (iv) Start of hit in full-length sequence. (v) End of 
**  hit in full-length sequence.  (vi) Whether a hit from the validation file 
**  for the family corresponding to the signature was redundant or not.  For
**  hits with a primary classification of 'TRUE' (see 'Notes') it has a value 
**  of 'REDUNDANT', 'NON_REDUNDANT'.  For other hits 'NOT_APPLICABLE' is given.
**  '.' is given if the validation file was not provided.  (vii) The primary 
**  (objective) classification of the hit. For true hits (genuine relatives 
**  to the signature as listed in the scop families file) one of 'SEED', 
**  'HIT' or 'OTHER') is given. Otherwise, 'CROSS', 'FALSE'
**  or 'UNKNOWN' is given ('.' is given if a validation file was not 
**  provided). (viii) The secondary classification of the hit, either 'TRUE', 
**  'CROSS', 'FALSE' or 'UNKNOWN' ('.' is given if a scop families file was not 
**  provided). The classes of hits are defined in 'Notes'.  (ix) Score of 
**  sequence-signature match.  (x) p-value (see 'Notes'). (xi) E-value (see 
**  'Notes').
**  (8) XX - used for spacing.
**  (9) // - The file ends with a line containing '//' only.
** 
**  Excerpt from a signature hits file
** 
**  DE   Results of signature search
**  XX
**  CL   Alpha and beta proteins (a/b)
**  XX
**  FO   NAD(P)-binding Rossmann-fold domains
**  XX
**  SF   NAD(P)-binding Rossmann-fold domains
**  XX
**  FA   Lactate & malate dehydrogenases, N-terminal domain
**  XX
**  SI   51848
**  XX
**  HI  1     Q9YHX4    21   136  NOT_APPLICABLE UNKNOWN   UNKNOWN   54.7   0.000  0.000
**  HI  2     Q9W6G6    19   128  NOT_APPLICABLE UNKNOWN   UNKNOWN   34.2   0.000  0.000
**  HI  3     P23379    60   173  NOT_APPLICABLE FALSE     FALSE     24.3   0.000  0.000
**  HI  4     Q9XDS8    60   173  NOT_APPLICABLE FALSE     FALSE     13.1   0.000  0.000
**  HI  5     P54768    103  219  NOT_APPLICABLE FALSE     FALSE     11.5   0.000  0.000
**  HI  6     P22194    178  296  NOT_APPLICABLE UNKNOWN   UNKNOWN   11.1   0.000  0.000
**  //
** 
**
**  Excerpts from an alignment file are shown (below). The records used 
**  are are as follows:
**  (1) The DE, CL, FO, SF, FA, SI, XX and // records have the same meaning as 
**  in the hits file (above).
**  (2) Other lines contain either a fragment of protein sequence preceeded
**  by an accession number, or a fragment of an alignment of a signature
**  to the protein sequence (signature positions are marked with a '*').
**  The two numbers on either side of the sequence are begin and end residue
**  numbers for that line.
** 
**  Excerpt from a signature alignment file
**
**  DE   Results of signature search
**  XX
**  CL   Alpha and beta proteins (a/b)
**  XX
**  FO   alpha/beta-Hydrolases
**  XX
**  SF   alpha/beta-Hydrolases
**  XX
**  FA   Acetylcholinesterase-like
**  XX
**  SI   14982
**  XX
**  OPSD_HUMAN      1        MNGTEGPNFYVPFSNATGVVRSPFEYPQYYLAEPWQFSMLAAYMF 45      
**  SIGNATURE       -        ---------*------------*---------------*------   
**  OPSD_XENLA      1        MNGTEGPNFYVPMSNKTGVVRSPFDYPQYYLAEPWQYSALAAYMF 45      
**  SIGNATURE       -        --------*-------------*----------------*-----   
**  XX
**  OPSD_HUMAN      46       LLIVLGFPINFLTLYVTVQHKKLRTPLNYILLNLAVADLFMVLGG 90
**  SIGNATURE       -        --------------*--------------*------------*--         
**  OPSD_XENLA      46       LLILLGLPINFMTLFVTIQHKKLRTPLNYILLNLVFANHFMVLCG 90      
**  SIGNATURE       -        --------------*--------------*------------*--         
**  XX
**  OPSD_HUMAN      91       FTSTLYTSLHGYFVFGPTGCNLEGFFATLGGEIALWSLVVLAIER 135     
**  SIGNATURE       -        ---------*--*--------------------------**----         
**  OPSD_XENLA      91       FTVTMYTSMHGYFIFGPTGCYIEGFFATLGGEVALWSLVVLAVER 135     
**  SIGNATURE       -        ---------*----*-------------------------**---         
**  XX1
**  //
**  
**  
**  
**  Data files
**  sigscan requires a residue substitution matrix.
**  
**  
**  
**  Diagnostic error messages
**  
**  
**  
**  Authors
**  Jon Ison (jison@hgmp.mrc.ac.uk)
**  
**  
**  
**  References
**  A key residues approach to the definition of protein families and analysis
**  of sparse family signatures.  JC Ison, AJ Bleasby, MJ Blades, SC Daniel, 
**  JH Parish, JBC Findlay.  PROTEINS: Structure, Function & Genetics.  2000, 
**  40:330-341
**
**  
**  
**  Alignment of a sparse protein signature with protein sequences: application
**  to fold prediction for three small globulins.  SC Daniel, JH Parish, 
**  JC Ison, MJ Blades & JBC Findlay.  FEBS Letters.  1999, 459:349-352. 
**
**  
**  
******************************************************************************/




#include "emboss.h"
#include <math.h>








/* @prog sigscan *************************************************************
**
** Scans a signature against swissprot and writes a signature hits file.
**
******************************************************************************/


int main(int argc, char **argv)
{
    AjPFile      sigin =NULL;        /*Signature input file*/
    AjPFile      targetf =NULL;      /*SCOP families file*/
    AjPFile      hitsf =NULL;        /*Hits output file*/
    AjPFile      alignf =NULL;       /*Alignment output file*/

    AjPSeqall    database=NULL;      /*Protein sequences to match signature 
				       against*/
    AjPMatrixf   sub  =NULL;         /*Residue substitution matrix*/
    float        gapo =0.0;          /*Gap insertion penalty*/
    float        gape =0.0;          /*Gap extension penalty*/

    ajint        nhits=0;            /*Max. no. of hits to write to output files*/
    ajint        nfalse=0;           /*Max. no. of hits to write to output files*/
    AjPHitlist   hits=NULL;          /*Hitlist to stores hits from 
				       signature-sequence matches */
    ajint        thresh=0;           /*Minimum length (residues) of overlap 
				       required for two hits with the same 
				       code to be counted as the same hit. */
    

    AjPSignature sig=NULL;           /*Signature data structure*/

    AjPHitlist   targets =NULL;	     /*Pointer to AjOHitlist structure */    
    AjPList      listtargets = NULL; /*List of AjOHitlist structures */    
    AjPStr       *nterm=NULL;        /*Holds N-terminal matching options from acd*/
    ajint         ntopt=0;           /*N-terminal option as int */
    ajint              x=0;   /* Housekeeping variable*/
    



    ajNamInit("emboss");
    ajAcdInitP("sigscan", argc, argv, "DOMAINATRIX");
    

    /* GET VALUES FROM ACD */
    sigin      = ajAcdGetInfile("sigin");
    database   = ajAcdGetSeqall("database");
    targetf    = ajAcdGetInfile("targetf");   
    sub        = ajAcdGetMatrixf("sub");
    gapo       = ajAcdGetFloat("gapo");
    gape       = ajAcdGetFloat("gape");
    nhits      = ajAcdGetInt("nhits");
    nfalse     = ajAcdGetInt("nfalse");
    thresh     = ajAcdGetInt("thresh");
    hitsf      = ajAcdGetOutfile("hitsf");
    alignf     = ajAcdGetOutfile("alignf"); 
    nterm      = ajAcdGetList("nterm");


    /*Assign N-terminal matching option */
    ajFmtScanS(nterm[0], "%d", &ntopt);
    
    
    /* READ SIGNATURE FILE */
    if(!ajXyzSignatureRead(sigin, &sig))
    {	
	ajMatrixfDel(&sub);
	ajFileClose(&sigin);
	ajFileClose(&targetf);
	ajFileClose(&hitsf);
	ajFileClose(&alignf);
	ajFatal("Error reading signature file");
    }   
    else
	ajFmtPrint("Signature file read ok\n");

    
    
    /* COMPILE SIGNATURE */
    if(!ajXyzSignatureCompile(&sig, gapo, gape, sub))
    {	
	ajXyzSignatureDel(&sig);
	ajMatrixfDel(&sub);
	ajFileClose(&sigin);
	ajFileClose(&targetf);
	ajFileClose(&hitsf);
	ajFileClose(&alignf);
	ajFatal("Error compiling signature");
    }  
    else
	ajFmtPrint("Signature compiled ok\n");


    /* READ SCOP FAMILIES FILE & FILL THE LIST OF TARGETS*/
    if(targetf)
    {
	listtargets = ajListNew();
	while(ajXyzHitlistRead(targetf,"//",&targets))
	    ajListPush(listtargets, (AjPHitlist) targets);
	ajFmtPrint("Targets read ok\n");
    }
    
    

    /* THIS CODE BLOCK IS FOR DEBUG ONLY */
    /*    printf("no. pos.: %d\n", sig->npos);
	  for(y=0;y<sig->npos;y++)
	  {
	  printf("POSITION %d   no. gaps: %d\n",y, sig->pos[y]->ngaps);
	  for(x=0;x<sig->pos[y]->ngaps; x++)
	  printf("GAP %d  SIZ %d  PEN %f\n", x, sig->pos[y]->gsiz[x], 
	  sig->pos[y]->gpen[x]);
	  printf("\n\nRES.SUB.VALS\n");
	  for(x=0; x<26;x++)
	  printf("%c %f\n", (char)('A'+x), sig->pos[y]->subs[x]);
	  }
	  */
	
    

    /* ALIGN THE SIGNATURE TO THE SEQUENCES & FILL Hitlist object */
    hits = ajXyzHitlistNew(0);
    ajXyzSignatureAlignSeqall(sig, database, nhits, &hits, ntopt);
    ajFmtPrint("Signature aligned to db ok\n");


    /* CLASSIFY THE HITS ACCORDING TO THE SCOP FAMILIES FILE */
    if(targetf)
    {
	ajXyzHitlistClassify(&hits, listtargets, thresh);
	ajFmtPrint("Hit classified ok\n");
    }
    else
    {
	for(x=0; x<hits->N; x++)
	{
	    ajStrAssC(&hits->hits[x]->Typeobj, "UNKNOWN");
	    ajStrAssC(&hits->hits[x]->Typesbj, "UNKNOWN");
	    ajStrAssC(&hits->hits[x]->Group, "NOT_APPLICABLE");
	}
    }
    

    
    /*WRITE HITS OUTPUT FILE */
    ajXyzSignatureHitsWrite(hitsf, sig, hits, nfalse);
    ajFmtPrint("Hits file written ok\n");


    /*WRITE HITS OUTPUT FILE */
  
    ajXyzSignatureAlignWriteBlock(alignf, sig, hits);
    ajFmtPrint("Alignments file written ok\n");


    /* TIDY UP AND EXIT */
    ajXyzHitlistDel(&hits);
    if(targetf)
    {
	while(ajListPop(listtargets, (void *) &targets))
	    ajXyzHitlistDel(&targets);
	ajListDel(&listtargets);
    }
    ajXyzSignatureDel(&sig);
    ajSeqallDel(&database);
    ajMatrixfDel(&sub);
    ajFileClose(&sigin);
    ajFileClose(&targetf);
    ajFileClose(&hitsf);
    ajFileClose(&alignf);

    ajExit();
    return 0;    
}
































